{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "7a4ad41c-1298-4f35-a6dd-bee7e6fd820c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# multiple transcripts are inherited from refseq as exon-chain duplicates\n",
    "# this can present an issue with assembly and quantification\n",
    "# we need to identify the dominant copy and remove duplicate"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "b94387c6-08b6-4ba7-852a-552bd590d3d8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# main imports\n",
    "import os\n",
    "import sys\n",
    "import csv\n",
    "import glob\n",
    "import math\n",
    "import shutil\n",
    "import random\n",
    "import importlib\n",
    "import subprocess\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "\n",
    "pd.set_option('display.max_columns', None)\n",
    "pd.set_option('display.max_colwidth', 50)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "a1eccd73-b6a3-441f-9069-430baa4d9bd7",
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 1\n",
    "\n",
    "sys.path.insert(0, \"/ccb/salz4-4/avaraby/orfanage/soft\")\n",
    "%aimport definitions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "26889daa-bc98-433e-8738-ee50b43038a4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# data\n",
    "ref_fasta = \"/home/avaraby1/genomes/human/hg38/hg38_p12_ucsc.fa\"\n",
    "grch_gtf = \"/ccb/salz8-1/avaraby/chess_maintenance_scripts/data/chess_github_clone/chess3.1.2.GRCh38.gtf\"\n",
    "chm_gtf = \"/ccb/salz8-1/avaraby/chess_maintenance_scripts/data/chess_github_clone/chess3.1.2.CHM13.gtf\"\n",
    "outbase = \"../tmp/chess3.1.2.fix_duplicated_tx\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "f05f85ee-26ab-4a77-a2e1-7103a9070cf9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# find duplicates\n",
    "# load exon chains and find those with the same chain\n",
    "# compare against intron-chain"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "bbd1f2b8-eb56-419f-8c0d-c164a375f9a0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# run borf on chess\n",
    "cmd = [\"/ccb/salz8-1/avaraby/chess_maintenance_scripts/soft/borf/borf/main.py\",\"--use_geneid\",\"--score\",\"-g\",grch_gtf,\"-o\",\"/ccb/salz8-1/avaraby/chess_maintenance_scripts/tmp/borf\"]\n",
    "subprocess.call(cmd)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "31597cd7-d783-4651-9533-51ebe2b4c773",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tid</th>\n",
       "      <th>has_cds</th>\n",
       "      <th>echain</th>\n",
       "      <th>ichain</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>mom</th>\n",
       "      <th>cchain</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CHS.1.1</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1+11874-12227,12613-12721,13221-14409</td>\n",
       "      <td>chr1+12227-12613,12721-13221</td>\n",
       "      <td>CHS.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CHS.2.1</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1-14362-14829,14970-15038,15796-15947,16607...</td>\n",
       "      <td>chr1-14829-14970,15038-15796,15947-16607,16765...</td>\n",
       "      <td>CHS.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CHS.166734.2</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1-17369-17391</td>\n",
       "      <td>-</td>\n",
       "      <td>CHS.166734</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CHS.166734.1</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1-17369-17436</td>\n",
       "      <td>-</td>\n",
       "      <td>CHS.166734</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CHS.166734.3</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1-17409-17431</td>\n",
       "      <td>-</td>\n",
       "      <td>CHS.166734</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>168446</th>\n",
       "      <td>CHS.57396.1</td>\n",
       "      <td>1</td>\n",
       "      <td>chrUn_KI270755v1+21574-26743,27337-27719</td>\n",
       "      <td>chrUn_KI270755v1+26743-27337</td>\n",
       "      <td>CHS.57396</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>168447</th>\n",
       "      <td>CHS.59138.3</td>\n",
       "      <td>1</td>\n",
       "      <td>chrX_KI270881v1_alt-65309-66585,69007-69092,69...</td>\n",
       "      <td>chrX_KI270881v1_alt-66585-69007,69092-69763,69...</td>\n",
       "      <td>CHS.59138</td>\n",
       "      <td>50.0</td>\n",
       "      <td>chrX_KI270881v1_alt-65972-66585,69007-69092,69...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>168448</th>\n",
       "      <td>CHS.59138.4</td>\n",
       "      <td>1</td>\n",
       "      <td>chrX_KI270881v1_alt-65309-66585,69007-69092,69...</td>\n",
       "      <td>chrX_KI270881v1_alt-66585-69007,69092-69763,69...</td>\n",
       "      <td>CHS.59138</td>\n",
       "      <td>50.0</td>\n",
       "      <td>chrX_KI270881v1_alt-65972-66585,69007-69092,69...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>168449</th>\n",
       "      <td>CHS.59138.1</td>\n",
       "      <td>1</td>\n",
       "      <td>chrX_KI270881v1_alt-65314-66585,69007-69092,69...</td>\n",
       "      <td>chrX_KI270881v1_alt-66585-69007,69092-69763,69...</td>\n",
       "      <td>CHS.59138</td>\n",
       "      <td>50.0</td>\n",
       "      <td>chrX_KI270881v1_alt-65972-66585,69007-69092,69...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>168450</th>\n",
       "      <td>CHS.59138.2</td>\n",
       "      <td>1</td>\n",
       "      <td>chrX_KI270881v1_alt-65314-66585,69007-69092,69...</td>\n",
       "      <td>chrX_KI270881v1_alt-66585-69007,69092-69763,69...</td>\n",
       "      <td>CHS.59138</td>\n",
       "      <td>50.0</td>\n",
       "      <td>chrX_KI270881v1_alt-65972-66585,69007-69092,69...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>168451 rows Ã— 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 tid  has_cds  \\\n",
       "0            CHS.1.1        1   \n",
       "1            CHS.2.1        1   \n",
       "2       CHS.166734.2        1   \n",
       "3       CHS.166734.1        1   \n",
       "4       CHS.166734.3        1   \n",
       "...              ...      ...   \n",
       "168446   CHS.57396.1        1   \n",
       "168447   CHS.59138.3        1   \n",
       "168448   CHS.59138.4        1   \n",
       "168449   CHS.59138.1        1   \n",
       "168450   CHS.59138.2        1   \n",
       "\n",
       "                                                   echain  \\\n",
       "0                chr1+11874-12227,12613-12721,13221-14409   \n",
       "1       chr1-14362-14829,14970-15038,15796-15947,16607...   \n",
       "2                                        chr1-17369-17391   \n",
       "3                                        chr1-17369-17436   \n",
       "4                                        chr1-17409-17431   \n",
       "...                                                   ...   \n",
       "168446           chrUn_KI270755v1+21574-26743,27337-27719   \n",
       "168447  chrX_KI270881v1_alt-65309-66585,69007-69092,69...   \n",
       "168448  chrX_KI270881v1_alt-65309-66585,69007-69092,69...   \n",
       "168449  chrX_KI270881v1_alt-65314-66585,69007-69092,69...   \n",
       "168450  chrX_KI270881v1_alt-65314-66585,69007-69092,69...   \n",
       "\n",
       "                                                   ichain     gene_id   mom  \\\n",
       "0                            chr1+12227-12613,12721-13221       CHS.1   NaN   \n",
       "1       chr1-14829-14970,15038-15796,15947-16607,16765...       CHS.2   NaN   \n",
       "2                                                       -  CHS.166734   NaN   \n",
       "3                                                       -  CHS.166734   NaN   \n",
       "4                                                       -  CHS.166734   NaN   \n",
       "...                                                   ...         ...   ...   \n",
       "168446                       chrUn_KI270755v1+26743-27337   CHS.57396   NaN   \n",
       "168447  chrX_KI270881v1_alt-66585-69007,69092-69763,69...   CHS.59138  50.0   \n",
       "168448  chrX_KI270881v1_alt-66585-69007,69092-69763,69...   CHS.59138  50.0   \n",
       "168449  chrX_KI270881v1_alt-66585-69007,69092-69763,69...   CHS.59138  50.0   \n",
       "168450  chrX_KI270881v1_alt-66585-69007,69092-69763,69...   CHS.59138  50.0   \n",
       "\n",
       "                                                   cchain  \n",
       "0                                                       -  \n",
       "1                                                       -  \n",
       "2                                                       -  \n",
       "3                                                       -  \n",
       "4                                                       -  \n",
       "...                                                   ...  \n",
       "168446                                                  -  \n",
       "168447  chrX_KI270881v1_alt-65972-66585,69007-69092,69...  \n",
       "168448  chrX_KI270881v1_alt-65972-66585,69007-69092,69...  \n",
       "168449  chrX_KI270881v1_alt-65972-66585,69007-69092,69...  \n",
       "168450  chrX_KI270881v1_alt-65972-66585,69007-69092,69...  \n",
       "\n",
       "[168451 rows x 7 columns]"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "c3 = definitions.get_chains(grch_gtf,\"exon\",True)\n",
    "c3[\"echain\"] = c3[\"seqid\"]+c3[\"strand\"]+c3.apply(lambda row: \",\".join(str(x[0])+\"-\"+str(x[1]) for x in row[\"chain\"]),axis=1)\n",
    "c3[\"ichain\"] = c3[\"seqid\"]+c3[\"strand\"]+c3.apply(lambda row: \",\".join(str(x[0])+\"-\"+str(x[1]) for x in definitions.chain_inv(row[\"chain\"])),axis=1)\n",
    "c3[\"ichain\"] = np.where(c3[\"ichain\"].str.endswith((\"-\",\"+\")),\"-\",c3[\"ichain\"])\n",
    "c3 = c3[[\"tid\",\"has_cds\",\"echain\",\"ichain\"]]\n",
    "\n",
    "c3 = c3.merge(definitions.get_attribute(grch_gtf,[\"gene_id\"]),on=\"tid\",how=\"left\")\n",
    "c3 = c3.merge(definitions.get_attribute(\"/ccb/salz8-1/avaraby/chess_maintenance_scripts/tmp/borf.gtf\",[\"mom\"]),on=\"tid\",how=\"left\")\n",
    "\n",
    "# attach cds chain to it as well\n",
    "c3_cds = definitions.get_chains(grch_gtf,\"CDS\",True)\n",
    "c3_cds[\"cchain\"] = c3_cds[\"seqid\"]+c3_cds[\"strand\"]+c3_cds.apply(lambda row: \",\".join(str(x[0])+\"-\"+str(x[1]) for x in row[\"chain\"]),axis=1)\n",
    "c3_cds = c3_cds[[\"tid\",\"cchain\"]]\n",
    "c3_cds[\"cchain\"] = np.where(c3_cds[\"cchain\"].str.endswith((\"-\",\"+\")),\"-\",c3_cds[\"cchain\"])\n",
    "\n",
    "# merge\n",
    "c3 = c3.merge(c3_cds,on=\"tid\")\n",
    "c3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "d535ed3b-dd77-4030-8ad7-49340b812bf1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tid</th>\n",
       "      <th>source</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>tid_mane</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>CHS.131169.0</td>\n",
       "      <td>MANE</td>\n",
       "      <td>CHS.131169</td>\n",
       "      <td>CHS.131169.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>CHS.15.1</td>\n",
       "      <td>MANE</td>\n",
       "      <td>CHS.15</td>\n",
       "      <td>CHS.15.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>CHS.20.5</td>\n",
       "      <td>MANE</td>\n",
       "      <td>CHS.20</td>\n",
       "      <td>CHS.20.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>80</th>\n",
       "      <td>CHS.39.2</td>\n",
       "      <td>CHESS</td>\n",
       "      <td>CHS.39</td>\n",
       "      <td>CHS.39.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>81</th>\n",
       "      <td>CHS.39.5</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>CHS.39</td>\n",
       "      <td>CHS.39.7</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             tid  source     gene_id      tid_mane\n",
       "11  CHS.131169.0    MANE  CHS.131169  CHS.131169.0\n",
       "27      CHS.15.1    MANE      CHS.15      CHS.15.1\n",
       "44      CHS.20.5    MANE      CHS.20      CHS.20.5\n",
       "80      CHS.39.2   CHESS      CHS.39      CHS.39.7\n",
       "81      CHS.39.5  RefSeq      CHS.39      CHS.39.7"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# compute similarity to mane and decide which ones to keep\n",
    "\n",
    "# for every gene, we need to get MANE transcript and CDS chain\n",
    "\n",
    "# load transcripts with source and gene_id\n",
    "mane_cmp_df = definitions.get_attribute(grch_gtf,[\"gene_id\"],cols=[1])\n",
    "mane_cmp_df.rename({1:\"source\"},axis=1,inplace=True)\n",
    "\n",
    "# split off mane and join on the dataframe to mark mane_tid for each transcript\n",
    "mane_cmp_df = mane_cmp_df.merge(mane_cmp_df[mane_cmp_df[\"source\"]==\"MANE\"][[\"tid\",\"gene_id\"]].rename({\"tid\":\"tid_mane\"},axis=1),on=\"gene_id\",how=\"left\")\n",
    "\n",
    "mane_cmp_df = mane_cmp_df[~(mane_cmp_df[\"tid_mane\"].isna())]\n",
    "\n",
    "mane_cmp_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "7194b670-6fb8-4a00-ab5d-4fef3ac687a0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tid</th>\n",
       "      <th>has_cds</th>\n",
       "      <th>seqid</th>\n",
       "      <th>strand</th>\n",
       "      <th>coords</th>\n",
       "      <th>chain</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CHS.1.1</td>\n",
       "      <td>0</td>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:11874-14409</td>\n",
       "      <td>()</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CHS.2.1</td>\n",
       "      <td>0</td>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>chr1:14362-29370</td>\n",
       "      <td>()</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CHS.166734.2</td>\n",
       "      <td>0</td>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>chr1:17369-17391</td>\n",
       "      <td>()</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CHS.166734.1</td>\n",
       "      <td>0</td>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>chr1:17369-17436</td>\n",
       "      <td>()</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CHS.166734.3</td>\n",
       "      <td>0</td>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>chr1:17409-17431</td>\n",
       "      <td>()</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            tid  has_cds seqid strand            coords chain\n",
       "0       CHS.1.1        0  chr1      +  chr1:11874-14409    ()\n",
       "1       CHS.2.1        0  chr1      -  chr1:14362-29370    ()\n",
       "2  CHS.166734.2        0  chr1      -  chr1:17369-17391    ()\n",
       "3  CHS.166734.1        0  chr1      -  chr1:17369-17436    ()\n",
       "4  CHS.166734.3        0  chr1      -  chr1:17409-17431    ()"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# load chains for all transcripts\n",
    "cds_chains = definitions.get_chains(grch_gtf,\"CDS\",True)\n",
    "cds_chains.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "3f029633-d65c-4bff-8684-5cb511e0e3ab",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tid_x</th>\n",
       "      <th>source</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>tid_mane</th>\n",
       "      <th>has_cds</th>\n",
       "      <th>seqid</th>\n",
       "      <th>strand</th>\n",
       "      <th>coords</th>\n",
       "      <th>chain</th>\n",
       "      <th>chain_mane</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CHS.131169.0</td>\n",
       "      <td>MANE</td>\n",
       "      <td>CHS.131169</td>\n",
       "      <td>CHS.131169.0</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:65419-71585</td>\n",
       "      <td>((65565, 65573), (69037, 70008))</td>\n",
       "      <td>((65565, 65573), (69037, 70008))</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CHS.15.1</td>\n",
       "      <td>MANE</td>\n",
       "      <td>CHS.15</td>\n",
       "      <td>CHS.15.1</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>chr1:450740-451678</td>\n",
       "      <td>((450740, 451678),)</td>\n",
       "      <td>((450740, 451678),)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CHS.20.5</td>\n",
       "      <td>MANE</td>\n",
       "      <td>CHS.20</td>\n",
       "      <td>CHS.20.5</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>chr1:685716-686654</td>\n",
       "      <td>((685716, 686654),)</td>\n",
       "      <td>((685716, 686654),)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CHS.39.2</td>\n",
       "      <td>CHESS</td>\n",
       "      <td>CHS.39</td>\n",
       "      <td>CHS.39.7</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:923923-944574</td>\n",
       "      <td>((924432, 924948), (925922, 926013), (930155, ...</td>\n",
       "      <td>((924432, 924948), (925922, 926013), (930155, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CHS.39.5</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>CHS.39</td>\n",
       "      <td>CHS.39.7</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:923923-944574</td>\n",
       "      <td>((924432, 924948), (925922, 926013), (930155, ...</td>\n",
       "      <td>((924432, 924948), (925922, 926013), (930155, ...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          tid_x  source     gene_id      tid_mane  has_cds seqid strand  \\\n",
       "0  CHS.131169.0    MANE  CHS.131169  CHS.131169.0        1  chr1      +   \n",
       "1      CHS.15.1    MANE      CHS.15      CHS.15.1        1  chr1      -   \n",
       "2      CHS.20.5    MANE      CHS.20      CHS.20.5        1  chr1      -   \n",
       "3      CHS.39.2   CHESS      CHS.39      CHS.39.7        1  chr1      +   \n",
       "4      CHS.39.5  RefSeq      CHS.39      CHS.39.7        1  chr1      +   \n",
       "\n",
       "               coords                                              chain  \\\n",
       "0    chr1:65419-71585                   ((65565, 65573), (69037, 70008))   \n",
       "1  chr1:450740-451678                                ((450740, 451678),)   \n",
       "2  chr1:685716-686654                                ((685716, 686654),)   \n",
       "3  chr1:923923-944574  ((924432, 924948), (925922, 926013), (930155, ...   \n",
       "4  chr1:923923-944574  ((924432, 924948), (925922, 926013), (930155, ...   \n",
       "\n",
       "                                          chain_mane  \n",
       "0                   ((65565, 65573), (69037, 70008))  \n",
       "1                                ((450740, 451678),)  \n",
       "2                                ((685716, 686654),)  \n",
       "3  ((924432, 924948), (925922, 926013), (930155, ...  \n",
       "4  ((924432, 924948), (925922, 926013), (930155, ...  "
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# merge onto the dataframe mane_cmp\n",
    "mane_cmp_df = mane_cmp_df.merge(cds_chains,on=\"tid\",how=\"left\")\n",
    "mane_cmp_df = mane_cmp_df.merge(cds_chains[[\"tid\",\"chain\"]].rename({\"chain\":\"chain_mane\"},axis=1),left_on=\"tid_mane\",right_on=\"tid\",how=\"left\")\n",
    "mane_cmp_df.drop(\"tid_y\",axis=1,inplace=True)\n",
    "mane_cmp_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "d6e0a068-2432-4cf5-b076-5f21bacecced",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tid</th>\n",
       "      <th>source</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>tid_mane</th>\n",
       "      <th>has_cds</th>\n",
       "      <th>seqid</th>\n",
       "      <th>strand</th>\n",
       "      <th>coords</th>\n",
       "      <th>chain</th>\n",
       "      <th>chain_mane</th>\n",
       "      <th>mod_chain</th>\n",
       "      <th>c1len</th>\n",
       "      <th>c2len</th>\n",
       "      <th>match_start</th>\n",
       "      <th>match_end</th>\n",
       "      <th>num_bp_extra</th>\n",
       "      <th>num_bp_missing</th>\n",
       "      <th>num_bp_inframe</th>\n",
       "      <th>num_bp_match</th>\n",
       "      <th>num_bp_outframe</th>\n",
       "      <th>lpi</th>\n",
       "      <th>ilpi</th>\n",
       "      <th>mlpi</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CHS.131169.0</td>\n",
       "      <td>MANE</td>\n",
       "      <td>CHS.131169</td>\n",
       "      <td>CHS.131169.0</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:65419-71585</td>\n",
       "      <td>((65565, 65573), (69037, 70008))</td>\n",
       "      <td>((65565, 65573), (69037, 70008))</td>\n",
       "      <td>[[65565, 65573, 0], [69037, 70008, 0]]</td>\n",
       "      <td>981</td>\n",
       "      <td>981</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>981</td>\n",
       "      <td>981</td>\n",
       "      <td>0</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CHS.15.1</td>\n",
       "      <td>MANE</td>\n",
       "      <td>CHS.15</td>\n",
       "      <td>CHS.15.1</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>chr1:450740-451678</td>\n",
       "      <td>((450740, 451678),)</td>\n",
       "      <td>((450740, 451678),)</td>\n",
       "      <td>[[450740, 451678, 0]]</td>\n",
       "      <td>939</td>\n",
       "      <td>939</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>939</td>\n",
       "      <td>939</td>\n",
       "      <td>0</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CHS.20.5</td>\n",
       "      <td>MANE</td>\n",
       "      <td>CHS.20</td>\n",
       "      <td>CHS.20.5</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>chr1:685716-686654</td>\n",
       "      <td>((685716, 686654),)</td>\n",
       "      <td>((685716, 686654),)</td>\n",
       "      <td>[[685716, 686654, 0]]</td>\n",
       "      <td>939</td>\n",
       "      <td>939</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>939</td>\n",
       "      <td>939</td>\n",
       "      <td>0</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CHS.39.2</td>\n",
       "      <td>CHESS</td>\n",
       "      <td>CHS.39</td>\n",
       "      <td>CHS.39.7</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:923923-944574</td>\n",
       "      <td>((924432, 924948), (925922, 926013), (930155, ...</td>\n",
       "      <td>((924432, 924948), (925922, 926013), (930155, ...</td>\n",
       "      <td>[[924432, 924948, 0], [925922, 926013, 0], [93...</td>\n",
       "      <td>2583</td>\n",
       "      <td>2535</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>48</td>\n",
       "      <td>0</td>\n",
       "      <td>2535</td>\n",
       "      <td>2535</td>\n",
       "      <td>0</td>\n",
       "      <td>100</td>\n",
       "      <td>98</td>\n",
       "      <td>98</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CHS.39.5</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>CHS.39</td>\n",
       "      <td>CHS.39.7</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:923923-944574</td>\n",
       "      <td>((924432, 924948), (925922, 926013), (930155, ...</td>\n",
       "      <td>((924432, 924948), (925922, 926013), (930155, ...</td>\n",
       "      <td>[[924432, 924948, 0], [925922, 926013, 0], [93...</td>\n",
       "      <td>2538</td>\n",
       "      <td>2535</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>2535</td>\n",
       "      <td>2535</td>\n",
       "      <td>0</td>\n",
       "      <td>100</td>\n",
       "      <td>99</td>\n",
       "      <td>99</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            tid  source     gene_id      tid_mane  has_cds seqid strand  \\\n",
       "0  CHS.131169.0    MANE  CHS.131169  CHS.131169.0        1  chr1      +   \n",
       "1      CHS.15.1    MANE      CHS.15      CHS.15.1        1  chr1      -   \n",
       "2      CHS.20.5    MANE      CHS.20      CHS.20.5        1  chr1      -   \n",
       "3      CHS.39.2   CHESS      CHS.39      CHS.39.7        1  chr1      +   \n",
       "4      CHS.39.5  RefSeq      CHS.39      CHS.39.7        1  chr1      +   \n",
       "\n",
       "               coords                                              chain  \\\n",
       "0    chr1:65419-71585                   ((65565, 65573), (69037, 70008))   \n",
       "1  chr1:450740-451678                                ((450740, 451678),)   \n",
       "2  chr1:685716-686654                                ((685716, 686654),)   \n",
       "3  chr1:923923-944574  ((924432, 924948), (925922, 926013), (930155, ...   \n",
       "4  chr1:923923-944574  ((924432, 924948), (925922, 926013), (930155, ...   \n",
       "\n",
       "                                          chain_mane  \\\n",
       "0                   ((65565, 65573), (69037, 70008))   \n",
       "1                                ((450740, 451678),)   \n",
       "2                                ((685716, 686654),)   \n",
       "3  ((924432, 924948), (925922, 926013), (930155, ...   \n",
       "4  ((924432, 924948), (925922, 926013), (930155, ...   \n",
       "\n",
       "                                           mod_chain  c1len  c2len  \\\n",
       "0             [[65565, 65573, 0], [69037, 70008, 0]]    981    981   \n",
       "1                              [[450740, 451678, 0]]    939    939   \n",
       "2                              [[685716, 686654, 0]]    939    939   \n",
       "3  [[924432, 924948, 0], [925922, 926013, 0], [93...   2583   2535   \n",
       "4  [[924432, 924948, 0], [925922, 926013, 0], [93...   2538   2535   \n",
       "\n",
       "   match_start  match_end  num_bp_extra  num_bp_missing  num_bp_inframe  \\\n",
       "0         True       True             0               0             981   \n",
       "1         True       True             0               0             939   \n",
       "2         True       True             0               0             939   \n",
       "3         True       True            48               0            2535   \n",
       "4         True       True             3               0            2535   \n",
       "\n",
       "   num_bp_match  num_bp_outframe  lpi  ilpi  mlpi  \n",
       "0           981                0  100   100   100  \n",
       "1           939                0  100   100   100  \n",
       "2           939                0  100   100   100  \n",
       "3          2535                0  100    98    98  \n",
       "4          2535                0  100    99    99  "
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mane_cmp_df[[\"mod_chain\",\"c1len\",\"c2len\",\"match_start\",\"match_end\",\"num_bp_extra\",\"num_bp_missing\",\"num_bp_inframe\",\"num_bp_match\",\"num_bp_outframe\",\"lpi\",\"ilpi\",\"mlpi\"]] = mane_cmp_df.apply(lambda row: definitions.compare_and_extract(row[\"chain\"],row[\"chain_mane\"],row[\"strand\"]),axis=1)\n",
    "mane_cmp_df.rename({\"tid_x\":\"tid\"},axis=1,inplace=True)\n",
    "mane_cmp_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "d58b1b98-abb1-4268-a825-343b92197209",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tid</th>\n",
       "      <th>has_cds</th>\n",
       "      <th>echain</th>\n",
       "      <th>ichain</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>mom</th>\n",
       "      <th>cchain</th>\n",
       "      <th>tid_mane</th>\n",
       "      <th>ilpi</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CHS.1.1</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1+11874-12227,12613-12721,13221-14409</td>\n",
       "      <td>chr1+12227-12613,12721-13221</td>\n",
       "      <td>CHS.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CHS.2.1</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1-14362-14829,14970-15038,15796-15947,16607...</td>\n",
       "      <td>chr1-14829-14970,15038-15796,15947-16607,16765...</td>\n",
       "      <td>CHS.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CHS.166734.2</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1-17369-17391</td>\n",
       "      <td>-</td>\n",
       "      <td>CHS.166734</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CHS.166734.1</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1-17369-17436</td>\n",
       "      <td>-</td>\n",
       "      <td>CHS.166734</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CHS.166734.3</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1-17409-17431</td>\n",
       "      <td>-</td>\n",
       "      <td>CHS.166734</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            tid  has_cds                                             echain  \\\n",
       "0       CHS.1.1        1           chr1+11874-12227,12613-12721,13221-14409   \n",
       "1       CHS.2.1        1  chr1-14362-14829,14970-15038,15796-15947,16607...   \n",
       "2  CHS.166734.2        1                                   chr1-17369-17391   \n",
       "3  CHS.166734.1        1                                   chr1-17369-17436   \n",
       "4  CHS.166734.3        1                                   chr1-17409-17431   \n",
       "\n",
       "                                              ichain     gene_id  mom cchain  \\\n",
       "0                       chr1+12227-12613,12721-13221       CHS.1  NaN      -   \n",
       "1  chr1-14829-14970,15038-15796,15947-16607,16765...       CHS.2  NaN      -   \n",
       "2                                                  -  CHS.166734  NaN      -   \n",
       "3                                                  -  CHS.166734  NaN      -   \n",
       "4                                                  -  CHS.166734  NaN      -   \n",
       "\n",
       "  tid_mane  ilpi  \n",
       "0      NaN   NaN  \n",
       "1      NaN   NaN  \n",
       "2      NaN   NaN  \n",
       "3      NaN   NaN  \n",
       "4      NaN   NaN  "
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "c3 = c3.merge(mane_cmp_df[[\"tid\",\"tid_mane\",\"ilpi\"]],on=\"tid\",how=\"left\")\n",
    "c3.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "68d4bad1-0679-43f3-9c39-a1cdbddf1c10",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# iterate over duplicates and decide what to do based on the comparisons\n",
    "\n",
    "# keep group information (all tids so we can add tag with duplicates)\n",
    "\n",
    "keep_tids = {}\n",
    "\n",
    "for (gid,echain), grp in c3.groupby(by=['gene_id','echain']):\n",
    "    if len(set(grp[\"tid\"].tolist()))==1:\n",
    "        keep_tids[grp[\"tid\"].tolist()[0]] = []\n",
    "    \n",
    "    # find one with highest ilpi to mane\n",
    "    max_ilpi = 0\n",
    "    max_ilpi_tid = None\n",
    "\n",
    "    max_mom = 0\n",
    "    max_mom_tid = None\n",
    "    for idx, row in grp.iterrows():\n",
    "        ilpi = 0\n",
    "        try:\n",
    "            ilpi = float(row[\"ilpi\"])\n",
    "        except:\n",
    "            ilpi = 0\n",
    "        if ilpi>max_ilpi:\n",
    "            max_ilpi = ilpi\n",
    "            max_ilpi_tid = row[\"tid\"]\n",
    "\n",
    "        mom = 0\n",
    "        try:\n",
    "            mom = float(row[\"mom\"])\n",
    "        except:\n",
    "            mom = 0\n",
    "        if mom>max_mom:\n",
    "            max_mom = mom\n",
    "            max_mom_tid = row[\"tid\"]\n",
    "\n",
    "    if max_ilpi > 0: # found match\n",
    "        keep_tids[max_ilpi_tid] = []\n",
    "        for idx, row in grp.iterrows():\n",
    "            keep_tids[max_ilpi_tid].append(row[\"tid\"])\n",
    "        # continue to next group\n",
    "        continue\n",
    "\n",
    "    if max_mom > 0: # found match\n",
    "        keep_tids[max_mom_tid] = []\n",
    "        for idx, row in grp.iterrows():\n",
    "            keep_tids[max_mom_tid].append(row[\"tid\"])\n",
    "        # continue to next group\n",
    "        continue\n",
    "\n",
    "    # for everything else - pick at random\n",
    "    rnd_tid = random.choice(grp[\"tid\"].tolist())\n",
    "    keep_tids[rnd_tid] = []\n",
    "    for idx, row in grp.iterrows():\n",
    "        keep_tids[rnd_tid].append(row[\"tid\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4c058a93-5aa4-4ac0-9d83-a397858fb2b6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# write out the output\n",
    "with open() as outFP:\n",
    "    with open() as inFP:\n",
    "        for line in inFP:\n",
    "            "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ba66fd7d-f9a8-4762-a867-ca0ee310e6a5",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b751df0d-4a34-423f-84cc-7693228908c0",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e4036da4-b5c7-4e40-bd63-09f90fe310bf",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
