{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "7a4ad41c-1298-4f35-a6dd-bee7e6fd820c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# multiple transcripts are inherited from refseq as exon-chain duplicates\n",
    "# this can present an issue with assembly and quantification\n",
    "# we need to identify the dominant copy and remove duplicate"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "b94387c6-08b6-4ba7-852a-552bd590d3d8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# main imports\n",
    "import os\n",
    "import sys\n",
    "import csv\n",
    "import glob\n",
    "import math\n",
    "import shutil\n",
    "import random\n",
    "import importlib\n",
    "import subprocess\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "\n",
    "pd.set_option('display.max_columns', None)\n",
    "pd.set_option('display.max_colwidth', 50)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "a1eccd73-b6a3-441f-9069-430baa4d9bd7",
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 1\n",
    "\n",
    "sys.path.insert(0, \"/ccb/salz4-4/avaraby/orfanage/soft\")\n",
    "%aimport definitions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "26889daa-bc98-433e-8738-ee50b43038a4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# data\n",
    "ref_fasta = \"/home/avaraby1/genomes/human/hg38/hg38_p12_ucsc.fa\"\n",
    "grch_gtf = \"/ccb/salz8-1/avaraby/chess_maintenance_scripts/data/chess_github_clone/chess3.1.2.GRCh38.gtf\"\n",
    "chm_gtf = \"/ccb/salz8-1/avaraby/chess_maintenance_scripts/data/chess_github_clone/chess3.1.2.CHM13.gtf\"\n",
    "mane_gtf = \"/ccb/salz8-1/avaraby/chess_maintenance_scripts/data/MANE.GRCh38.v1.3.refseq_genomic.gtf\"\n",
    "outbase = \"../tmp/chess3.1.2.fix_duplicated_tx\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "f05f85ee-26ab-4a77-a2e1-7103a9070cf9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# find duplicates\n",
    "# load exon chains and find those with the same chain\n",
    "# compare against intron-chain"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "b19a1f47-92ee-4bc6-bfec-0b0f0cd41802",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tid</th>\n",
       "      <th>has_cds</th>\n",
       "      <th>echain</th>\n",
       "      <th>ichain</th>\n",
       "      <th>cchain</th>\n",
       "      <th>gene_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>rna-NM_001005484.2</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1+65419-65433,65520-65573,69037-71585</td>\n",
       "      <td>chr1+65433-65520,65573-69037</td>\n",
       "      <td>chr1+65565-65573,69037-70008</td>\n",
       "      <td>OR4F5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rna-NM_001005221.2</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1-450740-451678</td>\n",
       "      <td>-</td>\n",
       "      <td>chr1-450740-451678</td>\n",
       "      <td>OR4F29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>rna-NM_001005277.1</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1-685716-686654</td>\n",
       "      <td>-</td>\n",
       "      <td>chr1-685716-686654</td>\n",
       "      <td>OR4F16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>rna-NM_001385641.1</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1+923923-924948,925922-926013,930155-930336...</td>\n",
       "      <td>chr1+924948-925922,926013-930155,930336-931039...</td>\n",
       "      <td>chr1+924432-924948,925922-926013,930155-930336...</td>\n",
       "      <td>SAMD11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>rna-NM_015658.4</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1-944203-944800,945057-945146,945518-945653...</td>\n",
       "      <td>chr1-944800-945057,945146-945518,945653-946173...</td>\n",
       "      <td>chr1-944694-944800,945057-945146,945518-945653...</td>\n",
       "      <td>NOC2L</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  tid  has_cds  \\\n",
       "0  rna-NM_001005484.2        1   \n",
       "1  rna-NM_001005221.2        1   \n",
       "2  rna-NM_001005277.1        1   \n",
       "3  rna-NM_001385641.1        1   \n",
       "4     rna-NM_015658.4        1   \n",
       "\n",
       "                                              echain  \\\n",
       "0           chr1+65419-65433,65520-65573,69037-71585   \n",
       "1                                 chr1-450740-451678   \n",
       "2                                 chr1-685716-686654   \n",
       "3  chr1+923923-924948,925922-926013,930155-930336...   \n",
       "4  chr1-944203-944800,945057-945146,945518-945653...   \n",
       "\n",
       "                                              ichain  \\\n",
       "0                       chr1+65433-65520,65573-69037   \n",
       "1                                                  -   \n",
       "2                                                  -   \n",
       "3  chr1+924948-925922,926013-930155,930336-931039...   \n",
       "4  chr1-944800-945057,945146-945518,945653-946173...   \n",
       "\n",
       "                                              cchain gene_name  \n",
       "0                       chr1+65565-65573,69037-70008     OR4F5  \n",
       "1                                 chr1-450740-451678    OR4F29  \n",
       "2                                 chr1-685716-686654    OR4F16  \n",
       "3  chr1+924432-924948,925922-926013,930155-930336...    SAMD11  \n",
       "4  chr1-944694-944800,945057-945146,945518-945653...     NOC2L  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mane = definitions.get_chains(mane_gtf,\"exon\",True)\n",
    "mane[\"echain\"] = mane[\"seqid\"]+mane[\"strand\"]+mane.apply(lambda row: \",\".join(str(x[0])+\"-\"+str(x[1]) for x in row[\"chain\"]),axis=1)\n",
    "mane[\"ichain\"] = mane[\"seqid\"]+mane[\"strand\"]+mane.apply(lambda row: \",\".join(str(x[0])+\"-\"+str(x[1]) for x in definitions.chain_inv(row[\"chain\"])),axis=1)\n",
    "mane[\"ichain\"] = np.where(mane[\"ichain\"].str.endswith((\"-\",\"+\")),\"-\",mane[\"ichain\"])\n",
    "mane = mane[[\"tid\",\"has_cds\",\"echain\",\"ichain\"]]\n",
    "\n",
    "# attach cds chain to it as well\n",
    "mane_cds = definitions.get_chains(mane_gtf,\"CDS\",True)\n",
    "mane_cds[\"cchain\"] = mane_cds[\"seqid\"]+mane_cds[\"strand\"]+mane_cds.apply(lambda row: \",\".join(str(x[0])+\"-\"+str(x[1]) for x in row[\"chain\"]),axis=1)\n",
    "mane_cds = mane_cds[[\"tid\",\"cchain\"]]\n",
    "\n",
    "# merge\n",
    "mane = mane.merge(mane_cds,on=\"tid\")\n",
    "\n",
    "# add gene names for ease of debugging\n",
    "gns = definitions.get_attribute(mane_gtf,\"gene_name\")\n",
    "mane = mane.merge(gns,on=\"tid\")\n",
    "mane.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "31597cd7-d783-4651-9533-51ebe2b4c773",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tid</th>\n",
       "      <th>has_cds</th>\n",
       "      <th>echain</th>\n",
       "      <th>ichain</th>\n",
       "      <th>cchain</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CHS.1.1</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1+11874-12227,12613-12721,13221-14409</td>\n",
       "      <td>chr1+12227-12613,12721-13221</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CHS.2.1</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1-14362-14829,14970-15038,15796-15947,16607...</td>\n",
       "      <td>chr1-14829-14970,15038-15796,15947-16607,16765...</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CHS.166734.2</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1-17369-17391</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CHS.166734.1</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1-17369-17436</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CHS.166734.3</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1-17409-17431</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>168446</th>\n",
       "      <td>CHS.57396.1</td>\n",
       "      <td>1</td>\n",
       "      <td>chrUn_KI270755v1+21574-26743,27337-27719</td>\n",
       "      <td>chrUn_KI270755v1+26743-27337</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>168447</th>\n",
       "      <td>CHS.59138.3</td>\n",
       "      <td>1</td>\n",
       "      <td>chrX_KI270881v1_alt-65309-66585,69007-69092,69...</td>\n",
       "      <td>chrX_KI270881v1_alt-66585-69007,69092-69763,69...</td>\n",
       "      <td>chrX_KI270881v1_alt-65972-66585,69007-69092,69...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>168448</th>\n",
       "      <td>CHS.59138.4</td>\n",
       "      <td>1</td>\n",
       "      <td>chrX_KI270881v1_alt-65309-66585,69007-69092,69...</td>\n",
       "      <td>chrX_KI270881v1_alt-66585-69007,69092-69763,69...</td>\n",
       "      <td>chrX_KI270881v1_alt-65972-66585,69007-69092,69...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>168449</th>\n",
       "      <td>CHS.59138.1</td>\n",
       "      <td>1</td>\n",
       "      <td>chrX_KI270881v1_alt-65314-66585,69007-69092,69...</td>\n",
       "      <td>chrX_KI270881v1_alt-66585-69007,69092-69763,69...</td>\n",
       "      <td>chrX_KI270881v1_alt-65972-66585,69007-69092,69...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>168450</th>\n",
       "      <td>CHS.59138.2</td>\n",
       "      <td>1</td>\n",
       "      <td>chrX_KI270881v1_alt-65314-66585,69007-69092,69...</td>\n",
       "      <td>chrX_KI270881v1_alt-66585-69007,69092-69763,69...</td>\n",
       "      <td>chrX_KI270881v1_alt-65972-66585,69007-69092,69...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>168451 rows Ã— 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 tid  has_cds  \\\n",
       "0            CHS.1.1        1   \n",
       "1            CHS.2.1        1   \n",
       "2       CHS.166734.2        1   \n",
       "3       CHS.166734.1        1   \n",
       "4       CHS.166734.3        1   \n",
       "...              ...      ...   \n",
       "168446   CHS.57396.1        1   \n",
       "168447   CHS.59138.3        1   \n",
       "168448   CHS.59138.4        1   \n",
       "168449   CHS.59138.1        1   \n",
       "168450   CHS.59138.2        1   \n",
       "\n",
       "                                                   echain  \\\n",
       "0                chr1+11874-12227,12613-12721,13221-14409   \n",
       "1       chr1-14362-14829,14970-15038,15796-15947,16607...   \n",
       "2                                        chr1-17369-17391   \n",
       "3                                        chr1-17369-17436   \n",
       "4                                        chr1-17409-17431   \n",
       "...                                                   ...   \n",
       "168446           chrUn_KI270755v1+21574-26743,27337-27719   \n",
       "168447  chrX_KI270881v1_alt-65309-66585,69007-69092,69...   \n",
       "168448  chrX_KI270881v1_alt-65309-66585,69007-69092,69...   \n",
       "168449  chrX_KI270881v1_alt-65314-66585,69007-69092,69...   \n",
       "168450  chrX_KI270881v1_alt-65314-66585,69007-69092,69...   \n",
       "\n",
       "                                                   ichain  \\\n",
       "0                            chr1+12227-12613,12721-13221   \n",
       "1       chr1-14829-14970,15038-15796,15947-16607,16765...   \n",
       "2                                                       -   \n",
       "3                                                       -   \n",
       "4                                                       -   \n",
       "...                                                   ...   \n",
       "168446                       chrUn_KI270755v1+26743-27337   \n",
       "168447  chrX_KI270881v1_alt-66585-69007,69092-69763,69...   \n",
       "168448  chrX_KI270881v1_alt-66585-69007,69092-69763,69...   \n",
       "168449  chrX_KI270881v1_alt-66585-69007,69092-69763,69...   \n",
       "168450  chrX_KI270881v1_alt-66585-69007,69092-69763,69...   \n",
       "\n",
       "                                                   cchain  \n",
       "0                                                       -  \n",
       "1                                                       -  \n",
       "2                                                       -  \n",
       "3                                                       -  \n",
       "4                                                       -  \n",
       "...                                                   ...  \n",
       "168446                                                  -  \n",
       "168447  chrX_KI270881v1_alt-65972-66585,69007-69092,69...  \n",
       "168448  chrX_KI270881v1_alt-65972-66585,69007-69092,69...  \n",
       "168449  chrX_KI270881v1_alt-65972-66585,69007-69092,69...  \n",
       "168450  chrX_KI270881v1_alt-65972-66585,69007-69092,69...  \n",
       "\n",
       "[168451 rows x 5 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "c3 = definitions.get_chains(grch_gtf,\"exon\",True)\n",
    "c3[\"echain\"] = c3[\"seqid\"]+c3[\"strand\"]+c3.apply(lambda row: \",\".join(str(x[0])+\"-\"+str(x[1]) for x in row[\"chain\"]),axis=1)\n",
    "c3[\"ichain\"] = c3[\"seqid\"]+c3[\"strand\"]+c3.apply(lambda row: \",\".join(str(x[0])+\"-\"+str(x[1]) for x in definitions.chain_inv(row[\"chain\"])),axis=1)\n",
    "c3[\"ichain\"] = np.where(c3[\"ichain\"].str.endswith((\"-\",\"+\")),\"-\",c3[\"ichain\"])\n",
    "c3 = c3[[\"tid\",\"has_cds\",\"echain\",\"ichain\"]]\n",
    "\n",
    "# attach cds chain to it as well\n",
    "c3_cds = definitions.get_chains(grch_gtf,\"CDS\",True)\n",
    "c3_cds[\"cchain\"] = c3_cds[\"seqid\"]+c3_cds[\"strand\"]+c3_cds.apply(lambda row: \",\".join(str(x[0])+\"-\"+str(x[1]) for x in row[\"chain\"]),axis=1)\n",
    "c3_cds = c3_cds[[\"tid\",\"cchain\"]]\n",
    "c3_cds[\"cchain\"] = np.where(c3_cds[\"cchain\"].str.endswith((\"-\",\"+\")),\"-\",c3_cds[\"cchain\"])\n",
    "\n",
    "# merge\n",
    "c3 = c3.merge(c3_cds,on=\"tid\")\n",
    "c3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "e7982478-eac6-437a-bcfa-82512971a3e0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "174\n",
      "12\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tid</th>\n",
       "      <th>has_cds</th>\n",
       "      <th>echain</th>\n",
       "      <th>ichain</th>\n",
       "      <th>cchain</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CHS.144560.1</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1+69879592-69879895</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CHS.167000.1</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1+69879592-69879895</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CHS.146088.1</td>\n",
       "      <td>1</td>\n",
       "      <td>chr2+97100587-97100874</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CHS.167737.1</td>\n",
       "      <td>1</td>\n",
       "      <td>chr2+97100587-97100874</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CHS.136513.14</td>\n",
       "      <td>1</td>\n",
       "      <td>chr4+9659025-9671347,9677311-9677934,9685007-9...</td>\n",
       "      <td>chr4+9671347-9677311,9677934-9685007,9685242-9...</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             tid  has_cds                                             echain  \\\n",
       "0   CHS.144560.1        1                             chr1+69879592-69879895   \n",
       "1   CHS.167000.1        1                             chr1+69879592-69879895   \n",
       "2   CHS.146088.1        1                             chr2+97100587-97100874   \n",
       "3   CHS.167737.1        1                             chr2+97100587-97100874   \n",
       "4  CHS.136513.14        1  chr4+9659025-9671347,9677311-9677934,9685007-9...   \n",
       "\n",
       "                                              ichain cchain  \n",
       "0                                                  -      -  \n",
       "1                                                  -      -  \n",
       "2                                                  -      -  \n",
       "3                                                  -      -  \n",
       "4  chr4+9671347-9677311,9677934-9685007,9685242-9...      -  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# isolate those without cds (where all duplicates are without cds\n",
    "\n",
    "keep_noncoding_tids = set()\n",
    "noncoding_duplicate_group_tids = []\n",
    "edup = c3[c3.duplicated(subset='echain', keep=False)].reset_index(drop=True)\n",
    "print(len(set(edup[\"echain\"])))\n",
    "for echain, grp in edup.groupby('echain'):\n",
    "    # if cds in mane - keep it - we can repeat duplicate detection afterwards\n",
    "    if list(set(grp[\"cchain\"].tolist())) == [\"-\"]:\n",
    "        noncoding_duplicate_group_tids.extend(grp[\"tid\"].tolist())\n",
    "        keep_noncoding_tids.add(grp.iloc[0][\"tid\"])\n",
    "\n",
    "noncoding_duplicates = c3[c3[\"tid\"].isin(noncoding_duplicate_group_tids)].reset_index(drop=True)\n",
    "print(len(keep_noncoding_tids))\n",
    "noncoding_duplicates.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "d4db6443-fc94-4cb3-8284-e0eef8070543",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tid</th>\n",
       "      <th>has_cds</th>\n",
       "      <th>echain</th>\n",
       "      <th>ichain</th>\n",
       "      <th>cchain</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CHS.131169.0</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1+65419-65433,65520-65573,69037-71585</td>\n",
       "      <td>chr1+65433-65520,65573-69037</td>\n",
       "      <td>chr1+65565-65573,69037-70008</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CHS.139257.1</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1-365134-365692,373144-373323,379769-379870...</td>\n",
       "      <td>chr1-365692-373144,373323-379769,379870-380897...</td>\n",
       "      <td>chr1-365555-365692,373144-373323,379769-379870...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CHS.15.1</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1-450740-451678</td>\n",
       "      <td>-</td>\n",
       "      <td>chr1-450740-451678</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CHS.18.2</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1-586287-586358,586821-586955,601398-601577...</td>\n",
       "      <td>chr1-586358-586821,586955-601398,601577-607955...</td>\n",
       "      <td>chr1-586839-586955,601398-601577,607955-608056...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CHS.20.5</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1-685716-686654</td>\n",
       "      <td>-</td>\n",
       "      <td>chr1-685716-686654</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            tid  has_cds                                             echain  \\\n",
       "0  CHS.131169.0        1           chr1+65419-65433,65520-65573,69037-71585   \n",
       "1  CHS.139257.1        1  chr1-365134-365692,373144-373323,379769-379870...   \n",
       "2      CHS.15.1        1                                 chr1-450740-451678   \n",
       "3      CHS.18.2        1  chr1-586287-586358,586821-586955,601398-601577...   \n",
       "4      CHS.20.5        1                                 chr1-685716-686654   \n",
       "\n",
       "                                              ichain  \\\n",
       "0                       chr1+65433-65520,65573-69037   \n",
       "1  chr1-365692-373144,373323-379769,379870-380897...   \n",
       "2                                                  -   \n",
       "3  chr1-586358-586821,586955-601398,601577-607955...   \n",
       "4                                                  -   \n",
       "\n",
       "                                              cchain  \n",
       "0                       chr1+65565-65573,69037-70008  \n",
       "1  chr1-365555-365692,373144-373323,379769-379870...  \n",
       "2                                 chr1-450740-451678  \n",
       "3  chr1-586839-586955,601398-601577,607955-608056...  \n",
       "4                                 chr1-685716-686654  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# lets get rid of all remaining non-coding transcripts since they are now guaranteed to exist alongside coding transcripts\n",
    "c3 = c3[~(c3[\"tid\"].isin(noncoding_duplicate_group_tids))].reset_index(drop=True)\n",
    "c3 = c3[~(c3[\"cchain\"]==\"-\")].reset_index(drop=True)\n",
    "c3.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "6f25d12e-6712-4119-b456-ce3e0dfbea1a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "162\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(56, 116)"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# iterate over and check with mane\n",
    "\n",
    "mane_matched_tids = []\n",
    "mane_group_tids = set() # all tids in groups where a transcript is matched to mane\n",
    "edup = c3[c3.duplicated(subset='echain', keep=False)].reset_index(drop=True)\n",
    "print(len(set(edup[\"echain\"])))\n",
    "mane_cchains = set(mane[\"cchain\"].tolist())\n",
    "for echain, grp in edup.groupby('echain'):\n",
    "    # if cds in mane - keep it - we can repeat duplicate detection afterwards\n",
    "    for idx, row in grp.iterrows():\n",
    "        if row[\"cchain\"] in mane_cchains:\n",
    "            mane_matched_tids.append(row[\"tid\"])\n",
    "            mane_group_tids.update(grp[\"tid\"].tolist())\n",
    "len(mane_matched_tids),len(mane_group_tids)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "d535ed3b-dd77-4030-8ad7-49340b812bf1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tid</th>\n",
       "      <th>source</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>tid_mane</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>CHS.131169.0</td>\n",
       "      <td>MANE</td>\n",
       "      <td>CHS.131169</td>\n",
       "      <td>CHS.131169.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>CHS.15.1</td>\n",
       "      <td>MANE</td>\n",
       "      <td>CHS.15</td>\n",
       "      <td>CHS.15.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>CHS.20.5</td>\n",
       "      <td>MANE</td>\n",
       "      <td>CHS.20</td>\n",
       "      <td>CHS.20.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>80</th>\n",
       "      <td>CHS.39.2</td>\n",
       "      <td>CHESS</td>\n",
       "      <td>CHS.39</td>\n",
       "      <td>CHS.39.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>81</th>\n",
       "      <td>CHS.39.5</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>CHS.39</td>\n",
       "      <td>CHS.39.7</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             tid  source     gene_id      tid_mane\n",
       "11  CHS.131169.0    MANE  CHS.131169  CHS.131169.0\n",
       "27      CHS.15.1    MANE      CHS.15      CHS.15.1\n",
       "44      CHS.20.5    MANE      CHS.20      CHS.20.5\n",
       "80      CHS.39.2   CHESS      CHS.39      CHS.39.7\n",
       "81      CHS.39.5  RefSeq      CHS.39      CHS.39.7"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# compute similarity to mane and decide which ones to keep\n",
    "\n",
    "# for every gene, we need to get MANE transcript and CDS chain\n",
    "\n",
    "# load transcripts with source and gene_id\n",
    "mane_cmp_df = definitions.get_attribute(grch_gtf,[\"gene_id\"],cols=[1])\n",
    "mane_cmp_df.rename({1:\"source\"},axis=1,inplace=True)\n",
    "\n",
    "# split off mane and join on the dataframe to mark mane_tid for each transcript\n",
    "mane_cmp_df = mane_cmp_df.merge(mane_cmp_df[mane_cmp_df[\"source\"]==\"MANE\"][[\"tid\",\"gene_id\"]].rename({\"tid\":\"tid_mane\"},axis=1),on=\"gene_id\",how=\"left\")\n",
    "\n",
    "mane_cmp_df = mane_cmp_df[~(mane_cmp_df[\"tid_mane\"].isna())]\n",
    "\n",
    "mane_cmp_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 156,
   "id": "7194b670-6fb8-4a00-ab5d-4fef3ac687a0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tid</th>\n",
       "      <th>has_cds</th>\n",
       "      <th>seqid</th>\n",
       "      <th>strand</th>\n",
       "      <th>coords</th>\n",
       "      <th>chain</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CHS.1.1</td>\n",
       "      <td>0</td>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:11874-14409</td>\n",
       "      <td>()</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CHS.2.1</td>\n",
       "      <td>0</td>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>chr1:14362-29370</td>\n",
       "      <td>()</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CHS.166734.2</td>\n",
       "      <td>0</td>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>chr1:17369-17391</td>\n",
       "      <td>()</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CHS.166734.1</td>\n",
       "      <td>0</td>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>chr1:17369-17436</td>\n",
       "      <td>()</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CHS.166734.3</td>\n",
       "      <td>0</td>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>chr1:17409-17431</td>\n",
       "      <td>()</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            tid  has_cds seqid strand            coords chain\n",
       "0       CHS.1.1        0  chr1      +  chr1:11874-14409    ()\n",
       "1       CHS.2.1        0  chr1      -  chr1:14362-29370    ()\n",
       "2  CHS.166734.2        0  chr1      -  chr1:17369-17391    ()\n",
       "3  CHS.166734.1        0  chr1      -  chr1:17369-17436    ()\n",
       "4  CHS.166734.3        0  chr1      -  chr1:17409-17431    ()"
      ]
     },
     "execution_count": 156,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# load chains for all transcripts\n",
    "cds_chains = definitions.get_chains(grch_gtf,\"CDS\",True)\n",
    "cds_chains.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 143,
   "id": "3f029633-d65c-4bff-8684-5cb511e0e3ab",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tid_x</th>\n",
       "      <th>source</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>tid_mane</th>\n",
       "      <th>has_cds</th>\n",
       "      <th>seqid</th>\n",
       "      <th>strand</th>\n",
       "      <th>coords</th>\n",
       "      <th>chain</th>\n",
       "      <th>chain_mane</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CHS.131169.0</td>\n",
       "      <td>MANE</td>\n",
       "      <td>CHS.131169</td>\n",
       "      <td>CHS.131169.0</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:65419-71585</td>\n",
       "      <td>((65565, 65573), (69037, 70008))</td>\n",
       "      <td>((65565, 65573), (69037, 70008))</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CHS.15.1</td>\n",
       "      <td>MANE</td>\n",
       "      <td>CHS.15</td>\n",
       "      <td>CHS.15.1</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>chr1:450740-451678</td>\n",
       "      <td>((450740, 451678),)</td>\n",
       "      <td>((450740, 451678),)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CHS.20.5</td>\n",
       "      <td>MANE</td>\n",
       "      <td>CHS.20</td>\n",
       "      <td>CHS.20.5</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>chr1:685716-686654</td>\n",
       "      <td>((685716, 686654),)</td>\n",
       "      <td>((685716, 686654),)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CHS.39.2</td>\n",
       "      <td>CHESS</td>\n",
       "      <td>CHS.39</td>\n",
       "      <td>CHS.39.7</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:923923-944574</td>\n",
       "      <td>((924432, 924948), (925922, 926013), (930155, ...</td>\n",
       "      <td>((924432, 924948), (925922, 926013), (930155, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CHS.39.5</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>CHS.39</td>\n",
       "      <td>CHS.39.7</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:923923-944574</td>\n",
       "      <td>((924432, 924948), (925922, 926013), (930155, ...</td>\n",
       "      <td>((924432, 924948), (925922, 926013), (930155, ...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          tid_x  source     gene_id      tid_mane  has_cds seqid strand  \\\n",
       "0  CHS.131169.0    MANE  CHS.131169  CHS.131169.0        1  chr1      +   \n",
       "1      CHS.15.1    MANE      CHS.15      CHS.15.1        1  chr1      -   \n",
       "2      CHS.20.5    MANE      CHS.20      CHS.20.5        1  chr1      -   \n",
       "3      CHS.39.2   CHESS      CHS.39      CHS.39.7        1  chr1      +   \n",
       "4      CHS.39.5  RefSeq      CHS.39      CHS.39.7        1  chr1      +   \n",
       "\n",
       "               coords                                              chain  \\\n",
       "0    chr1:65419-71585                   ((65565, 65573), (69037, 70008))   \n",
       "1  chr1:450740-451678                                ((450740, 451678),)   \n",
       "2  chr1:685716-686654                                ((685716, 686654),)   \n",
       "3  chr1:923923-944574  ((924432, 924948), (925922, 926013), (930155, ...   \n",
       "4  chr1:923923-944574  ((924432, 924948), (925922, 926013), (930155, ...   \n",
       "\n",
       "                                          chain_mane  \n",
       "0                   ((65565, 65573), (69037, 70008))  \n",
       "1                                ((450740, 451678),)  \n",
       "2                                ((685716, 686654),)  \n",
       "3  ((924432, 924948), (925922, 926013), (930155, ...  \n",
       "4  ((924432, 924948), (925922, 926013), (930155, ...  "
      ]
     },
     "execution_count": 143,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# merge onto the dataframe mane_cmp\n",
    "mane_cmp_df = mane_cmp_df.merge(cds_chains,on=\"tid\",how=\"left\")\n",
    "mane_cmp_df = mane_cmp_df.merge(cds_chains[[\"tid\",\"chain\"]].rename({\"chain\":\"chain_mane\"},axis=1),left_on=\"tid_mane\",right_on=\"tid\",how=\"left\")\n",
    "mane_cmp_df.drop(\"tid_y\",axis=1,inplace=True)\n",
    "mane_cmp_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 144,
   "id": "d6e0a068-2432-4cf5-b076-5f21bacecced",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tid</th>\n",
       "      <th>source</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>tid_mane</th>\n",
       "      <th>has_cds</th>\n",
       "      <th>seqid</th>\n",
       "      <th>strand</th>\n",
       "      <th>coords</th>\n",
       "      <th>chain</th>\n",
       "      <th>chain_mane</th>\n",
       "      <th>mod_chain</th>\n",
       "      <th>c1len</th>\n",
       "      <th>c2len</th>\n",
       "      <th>match_start</th>\n",
       "      <th>match_end</th>\n",
       "      <th>num_bp_extra</th>\n",
       "      <th>num_bp_missing</th>\n",
       "      <th>num_bp_inframe</th>\n",
       "      <th>num_bp_match</th>\n",
       "      <th>num_bp_outframe</th>\n",
       "      <th>lpi</th>\n",
       "      <th>ilpi</th>\n",
       "      <th>mlpi</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CHS.131169.0</td>\n",
       "      <td>MANE</td>\n",
       "      <td>CHS.131169</td>\n",
       "      <td>CHS.131169.0</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:65419-71585</td>\n",
       "      <td>((65565, 65573), (69037, 70008))</td>\n",
       "      <td>((65565, 65573), (69037, 70008))</td>\n",
       "      <td>[[65565, 65573, 0], [69037, 70008, 0]]</td>\n",
       "      <td>981</td>\n",
       "      <td>981</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>981</td>\n",
       "      <td>981</td>\n",
       "      <td>0</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CHS.15.1</td>\n",
       "      <td>MANE</td>\n",
       "      <td>CHS.15</td>\n",
       "      <td>CHS.15.1</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>chr1:450740-451678</td>\n",
       "      <td>((450740, 451678),)</td>\n",
       "      <td>((450740, 451678),)</td>\n",
       "      <td>[[450740, 451678, 0]]</td>\n",
       "      <td>939</td>\n",
       "      <td>939</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>939</td>\n",
       "      <td>939</td>\n",
       "      <td>0</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CHS.20.5</td>\n",
       "      <td>MANE</td>\n",
       "      <td>CHS.20</td>\n",
       "      <td>CHS.20.5</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>-</td>\n",
       "      <td>chr1:685716-686654</td>\n",
       "      <td>((685716, 686654),)</td>\n",
       "      <td>((685716, 686654),)</td>\n",
       "      <td>[[685716, 686654, 0]]</td>\n",
       "      <td>939</td>\n",
       "      <td>939</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>939</td>\n",
       "      <td>939</td>\n",
       "      <td>0</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CHS.39.2</td>\n",
       "      <td>CHESS</td>\n",
       "      <td>CHS.39</td>\n",
       "      <td>CHS.39.7</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:923923-944574</td>\n",
       "      <td>((924432, 924948), (925922, 926013), (930155, ...</td>\n",
       "      <td>((924432, 924948), (925922, 926013), (930155, ...</td>\n",
       "      <td>[[924432, 924948, 0], [925922, 926013, 0], [93...</td>\n",
       "      <td>2583</td>\n",
       "      <td>2535</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>48</td>\n",
       "      <td>0</td>\n",
       "      <td>2535</td>\n",
       "      <td>2535</td>\n",
       "      <td>0</td>\n",
       "      <td>100</td>\n",
       "      <td>98</td>\n",
       "      <td>98</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CHS.39.5</td>\n",
       "      <td>RefSeq</td>\n",
       "      <td>CHS.39</td>\n",
       "      <td>CHS.39.7</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>+</td>\n",
       "      <td>chr1:923923-944574</td>\n",
       "      <td>((924432, 924948), (925922, 926013), (930155, ...</td>\n",
       "      <td>((924432, 924948), (925922, 926013), (930155, ...</td>\n",
       "      <td>[[924432, 924948, 0], [925922, 926013, 0], [93...</td>\n",
       "      <td>2538</td>\n",
       "      <td>2535</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>2535</td>\n",
       "      <td>2535</td>\n",
       "      <td>0</td>\n",
       "      <td>100</td>\n",
       "      <td>99</td>\n",
       "      <td>99</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            tid  source     gene_id      tid_mane  has_cds seqid strand  \\\n",
       "0  CHS.131169.0    MANE  CHS.131169  CHS.131169.0        1  chr1      +   \n",
       "1      CHS.15.1    MANE      CHS.15      CHS.15.1        1  chr1      -   \n",
       "2      CHS.20.5    MANE      CHS.20      CHS.20.5        1  chr1      -   \n",
       "3      CHS.39.2   CHESS      CHS.39      CHS.39.7        1  chr1      +   \n",
       "4      CHS.39.5  RefSeq      CHS.39      CHS.39.7        1  chr1      +   \n",
       "\n",
       "               coords                                              chain  \\\n",
       "0    chr1:65419-71585                   ((65565, 65573), (69037, 70008))   \n",
       "1  chr1:450740-451678                                ((450740, 451678),)   \n",
       "2  chr1:685716-686654                                ((685716, 686654),)   \n",
       "3  chr1:923923-944574  ((924432, 924948), (925922, 926013), (930155, ...   \n",
       "4  chr1:923923-944574  ((924432, 924948), (925922, 926013), (930155, ...   \n",
       "\n",
       "                                          chain_mane  \\\n",
       "0                   ((65565, 65573), (69037, 70008))   \n",
       "1                                ((450740, 451678),)   \n",
       "2                                ((685716, 686654),)   \n",
       "3  ((924432, 924948), (925922, 926013), (930155, ...   \n",
       "4  ((924432, 924948), (925922, 926013), (930155, ...   \n",
       "\n",
       "                                           mod_chain  c1len  c2len  \\\n",
       "0             [[65565, 65573, 0], [69037, 70008, 0]]    981    981   \n",
       "1                              [[450740, 451678, 0]]    939    939   \n",
       "2                              [[685716, 686654, 0]]    939    939   \n",
       "3  [[924432, 924948, 0], [925922, 926013, 0], [93...   2583   2535   \n",
       "4  [[924432, 924948, 0], [925922, 926013, 0], [93...   2538   2535   \n",
       "\n",
       "   match_start  match_end  num_bp_extra  num_bp_missing  num_bp_inframe  \\\n",
       "0         True       True             0               0             981   \n",
       "1         True       True             0               0             939   \n",
       "2         True       True             0               0             939   \n",
       "3         True       True            48               0            2535   \n",
       "4         True       True             3               0            2535   \n",
       "\n",
       "   num_bp_match  num_bp_outframe  lpi  ilpi  mlpi  \n",
       "0           981                0  100   100   100  \n",
       "1           939                0  100   100   100  \n",
       "2           939                0  100   100   100  \n",
       "3          2535                0  100    98    98  \n",
       "4          2535                0  100    99    99  "
      ]
     },
     "execution_count": 144,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mane_cmp_df[[\"mod_chain\",\"c1len\",\"c2len\",\"match_start\",\"match_end\",\"num_bp_extra\",\"num_bp_missing\",\"num_bp_inframe\",\"num_bp_match\",\"num_bp_outframe\",\"lpi\",\"ilpi\",\"mlpi\"]] = mane_cmp_df.apply(lambda row: definitions.compare_and_extract(row[\"chain\"],row[\"chain_mane\"],row[\"strand\"]),axis=1)\n",
    "mane_cmp_df.rename({\"tid_x\":\"tid\"},axis=1,inplace=True)\n",
    "mane_cmp_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 146,
   "id": "d58b1b98-abb1-4268-a825-343b92197209",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tid</th>\n",
       "      <th>has_cds</th>\n",
       "      <th>echain</th>\n",
       "      <th>ichain</th>\n",
       "      <th>cchain</th>\n",
       "      <th>tid_mane</th>\n",
       "      <th>ilpi</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CHS.131169.0</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1+65419-65433,65520-65573,69037-71585</td>\n",
       "      <td>chr1+65433-65520,65573-69037</td>\n",
       "      <td>chr1+65565-65573,69037-70008</td>\n",
       "      <td>CHS.131169.0</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CHS.139257.1</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1-365134-365692,373144-373323,379769-379870...</td>\n",
       "      <td>chr1-365692-373144,373323-379769,379870-380897...</td>\n",
       "      <td>chr1-365555-365692,373144-373323,379769-379870...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CHS.15.1</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1-450740-451678</td>\n",
       "      <td>-</td>\n",
       "      <td>chr1-450740-451678</td>\n",
       "      <td>CHS.15.1</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CHS.18.2</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1-586287-586358,586821-586955,601398-601577...</td>\n",
       "      <td>chr1-586358-586821,586955-601398,601577-607955...</td>\n",
       "      <td>chr1-586839-586955,601398-601577,607955-608056...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CHS.20.5</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1-685716-686654</td>\n",
       "      <td>-</td>\n",
       "      <td>chr1-685716-686654</td>\n",
       "      <td>CHS.20.5</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            tid  has_cds                                             echain  \\\n",
       "0  CHS.131169.0        1           chr1+65419-65433,65520-65573,69037-71585   \n",
       "1  CHS.139257.1        1  chr1-365134-365692,373144-373323,379769-379870...   \n",
       "2      CHS.15.1        1                                 chr1-450740-451678   \n",
       "3      CHS.18.2        1  chr1-586287-586358,586821-586955,601398-601577...   \n",
       "4      CHS.20.5        1                                 chr1-685716-686654   \n",
       "\n",
       "                                              ichain  \\\n",
       "0                       chr1+65433-65520,65573-69037   \n",
       "1  chr1-365692-373144,373323-379769,379870-380897...   \n",
       "2                                                  -   \n",
       "3  chr1-586358-586821,586955-601398,601577-607955...   \n",
       "4                                                  -   \n",
       "\n",
       "                                              cchain      tid_mane   ilpi  \n",
       "0                       chr1+65565-65573,69037-70008  CHS.131169.0  100.0  \n",
       "1  chr1-365555-365692,373144-373323,379769-379870...           NaN    NaN  \n",
       "2                                 chr1-450740-451678      CHS.15.1  100.0  \n",
       "3  chr1-586839-586955,601398-601577,607955-608056...           NaN    NaN  \n",
       "4                                 chr1-685716-686654      CHS.20.5  100.0  "
      ]
     },
     "execution_count": 146,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "c3 = c3.merge(mane_cmp_df[[\"tid\",\"gene_id\",\"tid_mane\",\"ilpi\"]],on=\"tid\",how=\"left\")\n",
    "c3.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 149,
   "id": "68d4bad1-0679-43f3-9c39-a1cdbddf1c10",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['CHS.7654.6', 'CHS.7654.7']\n",
      "['CHS.175543.1', 'CHS.175543.2']\n",
      "['CHS.175545.1', 'CHS.175545.2']\n",
      "['CHS.175560.1', 'CHS.175560.2']\n",
      "['CHS.175566.1', 'CHS.175566.2']\n",
      "['CHS.175684.1', 'CHS.175684.2']\n",
      "['CHS.175688.10', 'CHS.175688.9']\n",
      "['CHS.175688.3', 'CHS.175688.4']\n",
      "['CHS.18537.16', 'CHS.18537.17']\n",
      "['CHS.175891.2', 'CHS.175891.3']\n",
      "['CHS.175891.4', 'CHS.175891.5']\n",
      "['CHS.18634.3', 'CHS.18634.4', 'CHS.18634.5']\n",
      "['CHS.18634.10', 'CHS.18634.11', 'CHS.18634.6', 'CHS.18634.7', 'CHS.18634.8', 'CHS.18634.9']\n",
      "['CHS.175910.1', 'CHS.175910.2']\n",
      "['CHS.25096.10', 'CHS.25096.9']\n"
     ]
    },
    {
     "ename": "AssertionError",
     "evalue": "multiple mane matches:                tid  has_cds  \\\n87182  CHS.26116.2        1   \n87183  CHS.59390.1        1   \n\n                                                  echain  \\\n87182  chr19-18868545-18869390,18869983-18870619,1887...   \n87183  chr19-18868545-18869390,18869983-18870619,1887...   \n\n                                                  ichain  \\\n87182  chr19-18869390-18869983,18870619-18878930,1887...   \n87183  chr19-18869390-18869983,18870619-18878930,1887...   \n\n                                                  cchain     tid_mane   ilpi  \n87182  chr19-18870577-18870619,18878930-18879039,1887...  CHS.26116.2  100.0  \n87183          chr19-18868597-18869390,18869983-18870307  CHS.59390.1  100.0  ",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAssertionError\u001b[0m                            Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[149], line 16\u001b[0m\n\u001b[1;32m     14\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m idx, row \u001b[38;5;129;01min\u001b[39;00m grp\u001b[38;5;241m.\u001b[39miterrows():\n\u001b[1;32m     15\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m row[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mcchain\u001b[39m\u001b[38;5;124m\"\u001b[39m] \u001b[38;5;129;01min\u001b[39;00m mane_cchains:\n\u001b[0;32m---> 16\u001b[0m         \u001b[38;5;28;01massert\u001b[39;00m matched_mane \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m \u001b[38;5;129;01mor\u001b[39;00m matched_mane\u001b[38;5;241m==\u001b[39mrow[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mtid\u001b[39m\u001b[38;5;124m\"\u001b[39m], \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mmultiple mane matches: \u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;241m+\u001b[39m\u001b[38;5;28mstr\u001b[39m(grp)\n\u001b[1;32m     17\u001b[0m         matched_mane \u001b[38;5;241m=\u001b[39m row[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mtid\u001b[39m\u001b[38;5;124m\"\u001b[39m]\n\u001b[1;32m     19\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m matched_mane \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m: \u001b[38;5;66;03m# found mane match\u001b[39;00m\n",
      "\u001b[0;31mAssertionError\u001b[0m: multiple mane matches:                tid  has_cds  \\\n87182  CHS.26116.2        1   \n87183  CHS.59390.1        1   \n\n                                                  echain  \\\n87182  chr19-18868545-18869390,18869983-18870619,1887...   \n87183  chr19-18868545-18869390,18869983-18870619,1887...   \n\n                                                  ichain  \\\n87182  chr19-18869390-18869983,18870619-18878930,1887...   \n87183  chr19-18869390-18869983,18870619-18878930,1887...   \n\n                                                  cchain     tid_mane   ilpi  \n87182  chr19-18870577-18870619,18878930-18879039,1887...  CHS.26116.2  100.0  \n87183          chr19-18868597-18869390,18869983-18870307  CHS.59390.1  100.0  "
     ]
    }
   ],
   "source": [
    "# iterate over duplicates and decide what to do based on the comparisons\n",
    "\n",
    "# keep group information (all tids so we can add tag with duplicates)\n",
    "\n",
    "mane_matched_tids = {} # dict mapping tid to keep/discard (1/0)\n",
    "mane_cmp_tids = {}\n",
    "\n",
    "mane_cchains = set(mane[\"cchain\"].tolist())\n",
    "for echain, grp in c3.groupby('echain'):\n",
    "    if len(grp)==1:\n",
    "        continue\n",
    "    # check if matching mane first\n",
    "    matched_mane = None\n",
    "    for idx, row in grp.iterrows():\n",
    "        if row[\"cchain\"] in mane_cchains:\n",
    "            assert matched_mane is None or matched_mane==row[\"tid\"], \"multiple mane matches: \"+str(grp)\n",
    "            matched_mane = row[\"tid\"]\n",
    "\n",
    "    if matched_mane is not None: # found mane match\n",
    "        for idx, row in grp.iterrows():\n",
    "            mane_matched_tids[row[\"tid\"]] = row[\"cchain\"] in mane_cchains\n",
    "        # continue to next group\n",
    "        continue\n",
    "\n",
    "    # for everything else - find one with highest ilpi to mane\n",
    "    max_ilpi = 0\n",
    "    for idx, row in grp.iterrows():\n",
    "        max_ilpi = max(max_ilpi,row[\"ilpi\"])\n",
    "\n",
    "    if max_ilpi > 0: # found match\n",
    "        for idx, row in grp.iterrows():\n",
    "            mane_cmp_tids[row[\"tid\"]] = row[\"ilpi\"] == max_ilpi\n",
    "        # continue to next group\n",
    "        continue\n",
    "\n",
    "    # for everything else - report\n",
    "    print(grp[\"tid\"].tolist())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 150,
   "id": "ba66fd7d-f9a8-4762-a867-ca0ee310e6a5",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "({'CHS.3309.1': True,\n",
       "  'CHS.3309.21': True,\n",
       "  'CHS.3309.19': True,\n",
       "  'CHS.657.3': True,\n",
       "  'CHS.657.6': False,\n",
       "  'CHS.4270.11': True,\n",
       "  'CHS.4270.102': True,\n",
       "  'CHS.4270.47': True,\n",
       "  'CHS.732.15': True,\n",
       "  'CHS.732.2': False,\n",
       "  'CHS.5175.11': False,\n",
       "  'CHS.5175.20': True,\n",
       "  'CHS.1196.15': False,\n",
       "  'CHS.1196.2': True,\n",
       "  'CHS.1547.2': True,\n",
       "  'CHS.1547.6': False,\n",
       "  'CHS.3530.18': False,\n",
       "  'CHS.3530.5': True,\n",
       "  'CHS.3784.11': True,\n",
       "  'CHS.3784.3': False,\n",
       "  'CHS.3784.1': False,\n",
       "  'CHS.3784.10': True,\n",
       "  'CHS.4689.2': False,\n",
       "  'CHS.4689.40': True,\n",
       "  'CHS.5646.2': True,\n",
       "  'CHS.5646.16': True,\n",
       "  'CHS.6165.6': True,\n",
       "  'CHS.6165.8': False,\n",
       "  'CHS.6381.4': True,\n",
       "  'CHS.6381.3': True,\n",
       "  'CHS.6381.5': True,\n",
       "  'CHS.6381.26': True,\n",
       "  'CHS.6381.17': True,\n",
       "  'CHS.6696.5': True,\n",
       "  'CHS.6696.16': True,\n",
       "  'CHS.6696.45': True,\n",
       "  'CHS.6696.28': True,\n",
       "  'CHS.6696.31': True,\n",
       "  'CHS.6163.19': True,\n",
       "  'CHS.6163.2': True,\n",
       "  'CHS.6163.17': True,\n",
       "  'CHS.6163.18': True,\n",
       "  'CHS.6163.13': True,\n",
       "  'CHS.6206.27': True,\n",
       "  'CHS.6206.17': True,\n",
       "  'CHS.9845.4': True,\n",
       "  'CHS.9845.5': True,\n",
       "  'CHS.8206.3': True,\n",
       "  'CHS.8206.4': True,\n",
       "  'CHS.8609.24': False,\n",
       "  'CHS.8609.7': True,\n",
       "  'CHS.8631.13': False,\n",
       "  'CHS.8631.6': True,\n",
       "  'CHS.7702.3': True,\n",
       "  'CHS.7702.6': True,\n",
       "  'CHS.7702.4': True,\n",
       "  'CHS.9430.5': True,\n",
       "  'CHS.9430.4': True,\n",
       "  'CHS.10662.40': True,\n",
       "  'CHS.10662.50': True,\n",
       "  'CHS.11364.2': True,\n",
       "  'CHS.11364.5': True,\n",
       "  'CHS.11636.3': True,\n",
       "  'CHS.11636.9': True,\n",
       "  'CHS.11636.6': True,\n",
       "  'CHS.10636.28': True,\n",
       "  'CHS.10636.30': True,\n",
       "  'CHS.12533.1': True,\n",
       "  'CHS.12533.3': True,\n",
       "  'CHS.12533.4': True,\n",
       "  'CHS.12533.2': True,\n",
       "  'CHS.11186.16': False,\n",
       "  'CHS.11186.5': True,\n",
       "  'CHS.11495.29': False,\n",
       "  'CHS.11495.5': True,\n",
       "  'CHS.16055.38': False,\n",
       "  'CHS.16055.8': True,\n",
       "  'CHS.17590.16': False,\n",
       "  'CHS.17590.5': True,\n",
       "  'CHS.19013.14': True,\n",
       "  'CHS.19013.28': True,\n",
       "  'CHS.19013.41': True,\n",
       "  'CHS.19195.5': True,\n",
       "  'CHS.19195.2': True,\n",
       "  'CHS.22852.2': True,\n",
       "  'CHS.22852.31': False,\n",
       "  'CHS.21179.1': False,\n",
       "  'CHS.21179.3': True,\n",
       "  'CHS.23357.2': True,\n",
       "  'CHS.23357.13': True,\n",
       "  'CHS.21105.1': True,\n",
       "  'CHS.21105.3': False,\n",
       "  'CHS.21217.11': False,\n",
       "  'CHS.21217.4': True,\n",
       "  'CHS.21217.12': False,\n",
       "  'CHS.21217.5': True,\n",
       "  'CHS.24177.2': True,\n",
       "  'CHS.24177.7': True,\n",
       "  'CHS.24177.6': True,\n",
       "  'CHS.24177.9': True,\n",
       "  'CHS.24312.8': True,\n",
       "  'CHS.24312.22': True,\n",
       "  'CHS.24360.17': False,\n",
       "  'CHS.24360.4': True,\n",
       "  'CHS.25686.1': True,\n",
       "  'CHS.25686.22': True,\n",
       "  'CHS.25686.2': True,\n",
       "  'CHS.25686.9': True,\n",
       "  'CHS.25686.20': True,\n",
       "  'CHS.25208.4': True,\n",
       "  'CHS.25208.2': True,\n",
       "  'CHS.25218.3': True,\n",
       "  'CHS.25218.13': True,\n",
       "  'CHS.25218.45': True,\n",
       "  'CHS.25218.4': True,\n",
       "  'CHS.26075.10': False,\n",
       "  'CHS.26075.2': True},\n",
       " {'CHS.3309.46': False,\n",
       "  'CHS.3309.48': True,\n",
       "  'CHS.3309.22': True,\n",
       "  'CHS.3309.23': True,\n",
       "  'CHS.3309.24': False,\n",
       "  'CHS.3309.25': True,\n",
       "  'CHS.4270.73': True,\n",
       "  'CHS.4270.14': False,\n",
       "  'CHS.4270.8': True,\n",
       "  'CHS.4270.10': True,\n",
       "  'CHS.4270.3': True,\n",
       "  'CHS.4270.4': True,\n",
       "  'CHS.4270.5': True,\n",
       "  'CHS.4270.72': True,\n",
       "  'CHS.4270.80': True,\n",
       "  'CHS.4270.82': True,\n",
       "  'CHS.4270.120': True,\n",
       "  'CHS.4270.1': True,\n",
       "  'CHS.4270.87': True,\n",
       "  'CHS.4270.90': True,\n",
       "  'CHS.4270.101': True,\n",
       "  'CHS.4270.113': True,\n",
       "  'CHS.3238.12': True,\n",
       "  'CHS.3238.33': False,\n",
       "  'CHS.5646.3': False,\n",
       "  'CHS.5646.4': False,\n",
       "  'CHS.5646.6': False,\n",
       "  'CHS.5646.7': False,\n",
       "  'CHS.5646.8': False,\n",
       "  'CHS.5646.11': True,\n",
       "  'CHS.5646.10': True,\n",
       "  'CHS.5646.20': True,\n",
       "  'CHS.5646.22': True,\n",
       "  'CHS.5646.21': True,\n",
       "  'CHS.6165.5': True,\n",
       "  'CHS.6165.7': False,\n",
       "  'CHS.6381.2': True,\n",
       "  'CHS.6381.1': False,\n",
       "  'CHS.6381.25': True,\n",
       "  'CHS.6381.8': True,\n",
       "  'CHS.6381.9': True,\n",
       "  'CHS.6381.18': True,\n",
       "  'CHS.6381.27': True,\n",
       "  'CHS.6696.2': False,\n",
       "  'CHS.6696.44': True,\n",
       "  'CHS.6696.19': False,\n",
       "  'CHS.6696.26': True,\n",
       "  'CHS.6696.20': False,\n",
       "  'CHS.6696.18': True,\n",
       "  'CHS.6696.48': True,\n",
       "  'CHS.6696.27': True,\n",
       "  'CHS.6696.29': False,\n",
       "  'CHS.6696.30': False,\n",
       "  'CHS.6696.32': True,\n",
       "  'CHS.6206.7': False,\n",
       "  'CHS.6206.3': False,\n",
       "  'CHS.6206.2': False,\n",
       "  'CHS.6206.4': False,\n",
       "  'CHS.6206.10': False,\n",
       "  'CHS.6206.11': False,\n",
       "  'CHS.6206.36': False,\n",
       "  'CHS.6206.19': True,\n",
       "  'CHS.6206.18': True,\n",
       "  'CHS.6206.13': True,\n",
       "  'CHS.6206.15': True,\n",
       "  'CHS.6206.12': True,\n",
       "  'CHS.6206.38': True,\n",
       "  'CHS.6206.16': True,\n",
       "  'CHS.6206.14': True,\n",
       "  'CHS.6206.34': False,\n",
       "  'CHS.9845.3': True,\n",
       "  'CHS.8206.25': False,\n",
       "  'CHS.8206.1': True,\n",
       "  'CHS.8206.2': True,\n",
       "  'CHS.7702.5': False,\n",
       "  'CHS.9430.6': True,\n",
       "  'CHS.9430.7': False,\n",
       "  'CHS.9430.3': False,\n",
       "  'CHS.9430.25': True,\n",
       "  'CHS.10662.33': False,\n",
       "  'CHS.10662.34': False,\n",
       "  'CHS.10662.35': True,\n",
       "  'CHS.10662.36': False,\n",
       "  'CHS.10662.60': False,\n",
       "  'CHS.10662.41': True,\n",
       "  'CHS.10662.42': True,\n",
       "  'CHS.10662.39': True,\n",
       "  'CHS.10662.61': True,\n",
       "  'CHS.10662.43': True,\n",
       "  'CHS.10662.44': True,\n",
       "  'CHS.10662.45': True,\n",
       "  'CHS.10662.46': True,\n",
       "  'CHS.10662.47': True,\n",
       "  'CHS.10662.52': True,\n",
       "  'CHS.10662.48': True,\n",
       "  'CHS.10662.49': True,\n",
       "  'CHS.10662.51': True,\n",
       "  'CHS.10662.80': True,\n",
       "  'CHS.10662.53': True,\n",
       "  'CHS.10662.54': True,\n",
       "  'CHS.10662.55': True,\n",
       "  'CHS.10662.62': True,\n",
       "  'CHS.10662.56': True,\n",
       "  'CHS.10662.57': True,\n",
       "  'CHS.10662.58': True,\n",
       "  'CHS.10662.59': True,\n",
       "  'CHS.10662.77': True,\n",
       "  'CHS.10662.92': True,\n",
       "  'CHS.10662.84': True,\n",
       "  'CHS.10662.78': True,\n",
       "  'CHS.11364.51': True,\n",
       "  'CHS.11364.52': True,\n",
       "  'CHS.11364.3': False,\n",
       "  'CHS.11364.15': False,\n",
       "  'CHS.11364.4': True,\n",
       "  'CHS.11364.6': True,\n",
       "  'CHS.11364.7': True,\n",
       "  'CHS.11364.8': True,\n",
       "  'CHS.11364.14': False,\n",
       "  'CHS.11636.10': True,\n",
       "  'CHS.11636.8': False,\n",
       "  'CHS.11636.25': False,\n",
       "  'CHS.10636.29': False,\n",
       "  'CHS.10636.4': True,\n",
       "  'CHS.10636.17': True,\n",
       "  'CHS.10636.19': True,\n",
       "  'CHS.10636.31': True,\n",
       "  'CHS.10636.20': True,\n",
       "  'CHS.10636.27': True,\n",
       "  'CHS.12533.25': True,\n",
       "  'CHS.16447.111': False,\n",
       "  'CHS.16447.42': True,\n",
       "  'CHS.16447.112': False,\n",
       "  'CHS.16447.54': True,\n",
       "  'CHS.19013.9': True,\n",
       "  'CHS.19013.15': False,\n",
       "  'CHS.19013.69': True,\n",
       "  'CHS.19013.11': True,\n",
       "  'CHS.19013.16': False,\n",
       "  'CHS.19013.12': False,\n",
       "  'CHS.19013.19': True,\n",
       "  'CHS.19013.21': True,\n",
       "  'CHS.19013.40': False,\n",
       "  'CHS.19013.48': False,\n",
       "  'CHS.19013.42': False,\n",
       "  'CHS.19013.43': True,\n",
       "  'CHS.19013.56': True,\n",
       "  'CHS.19013.55': True,\n",
       "  'CHS.19195.3': True,\n",
       "  'CHS.19195.4': False,\n",
       "  'CHS.23357.3': True,\n",
       "  'CHS.23357.10': True,\n",
       "  'CHS.23357.19': True,\n",
       "  'CHS.23357.21': True,\n",
       "  'CHS.23357.27': True,\n",
       "  'CHS.23357.30': True,\n",
       "  'CHS.23357.33': True,\n",
       "  'CHS.23357.35': True,\n",
       "  'CHS.23357.40': True,\n",
       "  'CHS.21217.13': False,\n",
       "  'CHS.21217.6': True,\n",
       "  'CHS.21217.16': False,\n",
       "  'CHS.21217.2': True,\n",
       "  'CHS.21217.14': False,\n",
       "  'CHS.21217.7': True,\n",
       "  'CHS.21217.17': False,\n",
       "  'CHS.21217.3': True,\n",
       "  'CHS.21217.1': True,\n",
       "  'CHS.21217.15': False,\n",
       "  'CHS.24177.19': True,\n",
       "  'CHS.24177.21': True,\n",
       "  'CHS.24177.18': True,\n",
       "  'CHS.24312.9': False,\n",
       "  'CHS.24312.34': False,\n",
       "  'CHS.24312.40': True,\n",
       "  'CHS.24312.23': True,\n",
       "  'CHS.24360.22': True,\n",
       "  'CHS.24360.23': False,\n",
       "  'CHS.24360.1': True,\n",
       "  'CHS.24360.21': False,\n",
       "  'CHS.25686.7': False,\n",
       "  'CHS.25686.40': False,\n",
       "  'CHS.25686.43': False,\n",
       "  'CHS.25686.47': False,\n",
       "  'CHS.25686.75': False,\n",
       "  'CHS.25686.42': True,\n",
       "  'CHS.25686.4': False,\n",
       "  'CHS.25686.39': True,\n",
       "  'CHS.25686.3': False,\n",
       "  'CHS.25686.51': False,\n",
       "  'CHS.25686.23': False,\n",
       "  'CHS.25686.27': False,\n",
       "  'CHS.25686.28': False,\n",
       "  'CHS.25686.30': False,\n",
       "  'CHS.25218.12': False,\n",
       "  'CHS.25218.41': True,\n",
       "  'CHS.25218.76': True,\n",
       "  'CHS.25218.10': True,\n",
       "  'CHS.25218.16': False,\n",
       "  'CHS.25218.60': False,\n",
       "  'CHS.25218.68': True,\n",
       "  'CHS.25218.73': True,\n",
       "  'CHS.25218.32': True})"
      ]
     },
     "execution_count": 150,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mane_matched_tids,mane_cmp_tids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b751df0d-4a34-423f-84cc-7693228908c0",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e4036da4-b5c7-4e40-bd63-09f90fe310bf",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
