{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "6bd5c71b-a593-4966-a7b1-51e7a55d0e23",
   "metadata": {},
   "outputs": [],
   "source": [
    "# main imports\n",
    "import os\n",
    "import sys\n",
    "import csv\n",
    "import glob\n",
    "import math\n",
    "import shutil\n",
    "import random\n",
    "import importlib\n",
    "import subprocess\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "\n",
    "pd.set_option('display.max_columns', None)\n",
    "pd.set_option('display.max_colwidth', 50)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "0b1e3a02-cde6-437d-b2ab-4b4b573dbd35",
   "metadata": {},
   "outputs": [],
   "source": [
    "# adjust CDS coordinates to include the stop codon"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "09a50177-9df8-40be-97c1-b77fd3de931a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Gffread solution:\n",
    "# --adj-stop flag: gffread -g ~/genomicData/hg38/hg38_p12_ucsc.fa --adj-stop -T -o tmp.adj.gtf tmp.gtf\n",
    "# for seleno-=proteins it will truncate them to the selenocysteine since it appears as a stop-codon\n",
    "# what if the stop codon is past the 3' end of the transcript? will gffread extend it nonetheless?\n",
    "#    Answer: partially. Gffread seems to have a bug when extending the coordinates.\n",
    "#    The CDS extends to the correct coordinate, but the exon-chain remains intact, leading to the CDS extending past the exon end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "5ed2c47a-fb62-4be6-bca1-7fea234cd28c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# custom solution\n",
    "# implement a python script to do this\n",
    "# extract protein-fasta file from the gtf/gff with gffread\n",
    "# load the transcripts from fasta indicating on whether the last base in stop codon or not\n",
    "# iterate over the gtf/gff looking up the stop codon status in the proteins dictionary\n",
    "# if missing stop - extend by 3 bases making sure to respect the exon coordinates\n",
    "#     can use the cut/intersect/etc methods from definitions\n",
    "# at the very end - rerun gffread to re-assign phase information\n",
    "\n",
    "# Additional considerations\n",
    "# for GTF need to process transcript,exon,cds at once, making sure if exon is extended so is the transcript\n",
    "# for GFF need to also consider gene records which might be tricky\n",
    "#    - alternatively we re-generate GFF after applying fixes to gtf\n",
    "# What if the next 3 bases are not a stop codon? What should we do then?\n",
    "#    After the initial run over the file applying fixes, we can extract proteins again and check. For anything that still has no stop codon - we should revert to the original version of the transcript instead and flag it as unchanged"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "4470972a-f46c-4fea-9833-45a28e085077",
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 1\n",
    "\n",
    "sys.path.insert(0, \"/ccb/salz4-4/avaraby/orfanage/soft\")\n",
    "%aimport definitions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "9e5d7dec-f841-4593-b65a-ee3053f5aa32",
   "metadata": {},
   "outputs": [],
   "source": [
    "# data\n",
    "ref_fasta = \"/home/avaraby1/genomes/human/chm13/chm13v2.0.fa\"\n",
    "input_gtf = \"/ccb/salz8-1/avaraby/chess_maintenance_scripts/tmp/chess3.0.1.CHM13.fix_dup_gid.gtf\"\n",
    "outbase = \"../tmp/chess3.0.1.CHM13.fix_dup_gid.adj_stop\"\n",
    "\n",
    "is_gff = definitions.gtf_or_gff(input_gtf)==\"gff\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "6ee27600-6391-43e7-b52b-42dc3482e432",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# run gffread to extract proteins first\n",
    "cmd = [\"gffread\",\"-g\",ref_fasta,\"-x\",outbase+\".cds_nt.fa\",input_gtf]\n",
    "subprocess.call(cmd)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "bca10f89-d6df-4f6b-b571-cca063bb0bce",
   "metadata": {},
   "outputs": [],
   "source": [
    "# load into a dictionary\n",
    "stop_status = definitions.load_fasta_dict(outbase+\".cds_nt.fa\")\n",
    "# iterate and replace sequence with True/False based on whether last character is \"*\"\n",
    "for tid,seq in stop_status.items():\n",
    "    if seq[-3:] in [\"TAG\",\"TAA\",\"TGA\"]:\n",
    "        stop_status[tid]=True\n",
    "    else:\n",
    "        stop_status[tid]=False"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "a29611c7-a4bf-42fa-9a11-7ff3b0665a9b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqid</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>attributes</th>\n",
       "      <th>tid</th>\n",
       "      <th>gid</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>exon_chain</th>\n",
       "      <th>cds_chain</th>\n",
       "      <th>has_cds</th>\n",
       "      <th>status</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr19</td>\n",
       "      <td>Liftoff</td>\n",
       "      <td>transcript</td>\n",
       "      <td>4460</td>\n",
       "      <td>6999</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"CHS.1.1\"; gene_id \"CHS.1\"; gene...</td>\n",
       "      <td>CHS.1.1</td>\n",
       "      <td>CHS.1</td>\n",
       "      <td>CHS.1</td>\n",
       "      <td>((4460, 4813), (5199, 5307), (5807, 6999))</td>\n",
       "      <td>()</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>Liftoff</td>\n",
       "      <td>transcript</td>\n",
       "      <td>864517</td>\n",
       "      <td>868933</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"CHS.100.1\"; gene_id \"CHS.100\"; ...</td>\n",
       "      <td>CHS.100.1</td>\n",
       "      <td>CHS.100</td>\n",
       "      <td>CHS.100</td>\n",
       "      <td>((864517, 865022), (868548, 868933))</td>\n",
       "      <td>()</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>Liftoff</td>\n",
       "      <td>transcript</td>\n",
       "      <td>864517</td>\n",
       "      <td>868933</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"CHS.100.2\"; gene_id \"CHS.100\"; ...</td>\n",
       "      <td>CHS.100.2</td>\n",
       "      <td>CHS.100</td>\n",
       "      <td>CHS.100</td>\n",
       "      <td>((864517, 865022), (868538, 868933))</td>\n",
       "      <td>()</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>Liftoff</td>\n",
       "      <td>transcript</td>\n",
       "      <td>864517</td>\n",
       "      <td>868933</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"CHS.100.3\"; gene_id \"CHS.100\"; ...</td>\n",
       "      <td>CHS.100.3</td>\n",
       "      <td>CHS.100</td>\n",
       "      <td>CHS.100</td>\n",
       "      <td>((864517, 865314), (868538, 868933))</td>\n",
       "      <td>()</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>CHESS</td>\n",
       "      <td>transcript</td>\n",
       "      <td>864517</td>\n",
       "      <td>869839</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"CHS.100.8\"; gene_id \"CHS.100\"; ...</td>\n",
       "      <td>CHS.100.8</td>\n",
       "      <td>CHS.100</td>\n",
       "      <td>CHS.100</td>\n",
       "      <td>((864517, 865314), (868538, 868944), (869739, ...</td>\n",
       "      <td>()</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   seqid   source        type   start     end score strand phase  \\\n",
       "0  chr19  Liftoff  transcript    4460    6999     .      +     .   \n",
       "1   chr1  Liftoff  transcript  864517  868933     .      -     .   \n",
       "2   chr1  Liftoff  transcript  864517  868933     .      -     .   \n",
       "3   chr1  Liftoff  transcript  864517  868933     .      -     .   \n",
       "4   chr1    CHESS  transcript  864517  869839     .      -     .   \n",
       "\n",
       "                                          attributes        tid      gid  \\\n",
       "0  transcript_id \"CHS.1.1\"; gene_id \"CHS.1\"; gene...    CHS.1.1    CHS.1   \n",
       "1  transcript_id \"CHS.100.1\"; gene_id \"CHS.100\"; ...  CHS.100.1  CHS.100   \n",
       "2  transcript_id \"CHS.100.2\"; gene_id \"CHS.100\"; ...  CHS.100.2  CHS.100   \n",
       "3  transcript_id \"CHS.100.3\"; gene_id \"CHS.100\"; ...  CHS.100.3  CHS.100   \n",
       "4  transcript_id \"CHS.100.8\"; gene_id \"CHS.100\"; ...  CHS.100.8  CHS.100   \n",
       "\n",
       "  gene_name                                         exon_chain cds_chain  \\\n",
       "0     CHS.1         ((4460, 4813), (5199, 5307), (5807, 6999))        ()   \n",
       "1   CHS.100               ((864517, 865022), (868548, 868933))        ()   \n",
       "2   CHS.100               ((864517, 865022), (868538, 868933))        ()   \n",
       "3   CHS.100               ((864517, 865314), (868538, 868933))        ()   \n",
       "4   CHS.100  ((864517, 865314), (868538, 868944), (869739, ...        ()   \n",
       "\n",
       "   has_cds status  \n",
       "0        0    NaN  \n",
       "1        0    NaN  \n",
       "2        0    NaN  \n",
       "3        0    NaN  \n",
       "4        0    NaN  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# load exon and cds chains\n",
    "\n",
    "# load gtf to get the attributes (keep only the transcript lines)\n",
    "df = definitions.load_gtf(input_gtf)\n",
    "df = df[df[\"type\"]==\"transcript\"].reset_index(drop=True)\n",
    "df[\"tid\"] = df[\"attributes\"].str.split(\"transcript_id \\\"\",expand=True)[1].str.split(\"\\\"\",expand=True)[0]\n",
    "df[\"gid\"] = df[\"attributes\"].str.split(\"gene_id \\\"\",expand=True)[1].str.split(\"\\\"\",expand=True)[0]\n",
    "df[\"gene_name\"] = df[\"attributes\"].str.split(\"gene_id \\\"\",expand=True)[1].str.split(\"\\\"\",expand=True)[0]\n",
    "# load and merge exon chains\n",
    "exons = definitions.get_chains(input_gtf,feature_type=\"exon\",coords=False,phase=False)\n",
    "df = df.merge(exons[[\"tid\",\"chain\"]],on=\"tid\",how=\"outer\",indicator=\"merge\")\n",
    "df.rename({\"chain\":\"exon_chain\"},axis=1,inplace=True)\n",
    "assert len(df[~(df[\"merge\"]==\"both\")])==0,\"invalid merge of exons\"\n",
    "df.drop(\"merge\",axis=1,inplace=True)\n",
    "# load and merge cds chains\n",
    "cds = definitions.get_chains(input_gtf,feature_type=\"CDS\",coords=True,phase=False)\n",
    "df = df.merge(cds[[\"tid\",\"chain\",\"has_cds\"]],on=\"tid\",how=\"outer\",indicator=\"merge\")\n",
    "df.rename({\"chain\":\"cds_chain\"},axis=1,inplace=True)\n",
    "assert len(df[~(df[\"merge\"]==\"both\")])==0,\"invalid merge of cds\"\n",
    "df.drop(\"merge\",axis=1,inplace=True)\n",
    "\n",
    "# add the stop_status info\n",
    "df['status'] = df['tid'].map(stop_status)\n",
    "\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "bec0499b-92f0-46d7-af32-3c573bf9c8c1",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_n(chain, n, forward=True):\n",
    "    res = []\n",
    "    count = 0\n",
    "    \n",
    "    if forward:\n",
    "        for start, end in chain:\n",
    "            ilen = end - start + 1\n",
    "            if count + ilen < n:\n",
    "                res.append((start, end))\n",
    "                count += ilen\n",
    "            else:\n",
    "                res.append((start, start + n - count - 1))\n",
    "                break\n",
    "    else:\n",
    "        for start, end in reversed(chain):\n",
    "            ilen = end - start + 1\n",
    "            if count + ilen < n:\n",
    "                res.insert(0, (start, end))\n",
    "                count += ilen\n",
    "            else:\n",
    "                res.insert(0, (end - (n - count - 1), end))\n",
    "                break\n",
    "\n",
    "    return res\n",
    "\n",
    "def extend_intervals(c1, c2, num_coords=3, forward=True):\n",
    "    if forward:\n",
    "        # first cut the first list to keep only the coordinates that are in the first list on the appropriate end\n",
    "        c1_cut = definitions.cut(c1,c2[-1][1],c1[-1][1])\n",
    "        # skip the first position because it includes the last position in first list\n",
    "        if c1_cut[0][0]==c1_cut[0][1]:\n",
    "            c1_cut.pop(0)\n",
    "        else:\n",
    "            tmp = list(c1_cut[0])\n",
    "            tmp[0]+=1\n",
    "            c1_cut[0] = tuple(tmp)\n",
    "\n",
    "        # extract first N positions from the cut chain\n",
    "        extracted_positions = get_n(c1_cut,num_coords)\n",
    "\n",
    "        # append these to the second chain\n",
    "        c2.extend(extracted_positions)\n",
    "        c2 = definitions.merge(c2,True)\n",
    "        c2 = [tuple(x) for x in c2]\n",
    "        \n",
    "        if definitions.clen(extracted_positions)==num_coords:\n",
    "            pass\n",
    "        elif definitions.clen(extracted_positions)<num_coords: # not long enough? Add missing pieces\n",
    "            for i in range(num_coords-definitions.clen(extracted_positions)):\n",
    "                tmp = list(c2[-1])\n",
    "                tmp[1]+=1\n",
    "                c2[-1] = tuple(tmp)\n",
    "\n",
    "                tmp = list(c1[-1])\n",
    "                tmp[1]+=1\n",
    "                c1[-1] = tuple(tmp)\n",
    "        else:\n",
    "            raise Exception(c1,c2,num_coords)\n",
    "\n",
    "        return c1,c2\n",
    "\n",
    "    else:\n",
    "        # first cut the first list to keep only the coordinates that are in the first list on the appropriate end\n",
    "        c1_cut = definitions.cut(c1,c1[0][0],c2[0][0])\n",
    "        # skip the last position because it includes the last position in first list\n",
    "        if c1_cut[-1][0]==c1_cut[-1][1]:\n",
    "            c1_cut.pop()\n",
    "        else:\n",
    "            tmp = list(c1_cut[-1])\n",
    "            tmp[1]-=1\n",
    "            c1_cut[-1] = tuple(tmp)\n",
    "\n",
    "        # extract last N positions from the cut chain\n",
    "        extracted_positions = get_n(c1_cut,num_coords,False)\n",
    "\n",
    "        # append these to the second chain\n",
    "        c2 = extracted_positions+c2\n",
    "        c2 = definitions.merge(c2,True)\n",
    "        c2 = [tuple(x) for x in c2]\n",
    "        \n",
    "        if definitions.clen(extracted_positions)==num_coords:\n",
    "            pass\n",
    "        elif definitions.clen(extracted_positions)<num_coords: # not long enough? Add missing pieces\n",
    "            for i in range(num_coords-definitions.clen(extracted_positions)):\n",
    "                tmp = list(c2[0])\n",
    "                tmp[0]-=1\n",
    "                c2[0] = tuple(tmp)\n",
    "\n",
    "                tmp = list(c1[0])\n",
    "                tmp[0]-=1\n",
    "                c1[0] = tuple(tmp)\n",
    "        else:\n",
    "            raise Exception(c1,c2,num_coords)\n",
    "\n",
    "        return c1,c2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "50538872-aeb7-485a-b2b1-c9bd4f43aa92",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# process line by line - remember strands\n",
    "def fix_coordinates(row):\n",
    "    if not row[\"has_cds\"]:\n",
    "        return [row[\"exon_chain\"], row[\"cds_chain\"]]\n",
    "    if not definitions.clen(row[\"cds_chain\"])%3==0: # invalid CDS\n",
    "        return [row[\"exon_chain\"], row[\"cds_chain\"]]\n",
    "    if not definitions.cut(row[\"exon_chain\"],row[\"cds_chain\"][0][0],row[\"cds_chain\"][-1][1])==list(row[\"cds_chain\"]): # if the exon cahin can not be cut to the cds chain\n",
    "        return [row[\"exon_chain\"], row[\"cds_chain\"]]\n",
    "    if row[\"status\"]==True: # already has stop\n",
    "        return [row[\"exon_chain\"], row[\"cds_chain\"]]\n",
    "    if row[\"exon_chain\"][0][0]<4: # too close to start\n",
    "        return [row[\"exon_chain\"], row[\"cds_chain\"]]\n",
    "        \n",
    "    # otherwise process the data\n",
    "    c1,c2 = extend_intervals(list(row[\"exon_chain\"]),list(row[\"cds_chain\"]),3,row[\"strand\"]==\"+\")\n",
    "    return [c1,c2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "1221802d-eff0-4e62-846a-ec00565ef72f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# now we need to iterate over the dataframe and write out the transcripts for the initial validation\n",
    "\n",
    "def to_gtf(row,outFP):\n",
    "    exons_adjusted = [tuple(x) for x in row[\"exon_chain\"]] != [tuple(x) for x in row[\"new_exon_chain\"]]\n",
    "    cds_adjusted = [tuple(x) for x in row[\"cds_chain\"]] != [tuple(x) for x in row[\"new_cds_chain\"]]\n",
    "    tlcs     = [row[\"seqid\"],\n",
    "                row[\"source\"],\n",
    "                row[\"type\"],\n",
    "                str(row[\"new_exon_chain\"][0][0]),\n",
    "                str(row[\"new_exon_chain\"][-1][1]),\n",
    "                row[\"score\"],\n",
    "                row[\"strand\"],\n",
    "                row[\"phase\"],\n",
    "                row[\"attributes\"].rstrip().rstrip(\";\").rstrip()+\"; exon_adjustment_status \\\"\"+str(int(exons_adjusted))+\"\\\"; cds_adjustment_status \\\"\"+str(int(cds_adjusted))+\"\\\";\"]\n",
    "    outFP.write(\"\\t\".join(tlcs)+\"\\n\")\n",
    "\n",
    "    \n",
    "    exon_attrs = definitions.extract_attributes(row[\"attributes\"])\n",
    "    allowed_keys = {\"transcript_id\", \"gene_id\", \"gene_name\"}\n",
    "    exon_attrs = {key: value for key, value in exon_attrs.items() if key in allowed_keys}\n",
    "    exon_attrs = definitions.to_attribute_string(exon_attrs)\n",
    "\n",
    "    for exon in row[\"new_exon_chain\"]:\n",
    "        elcs = [row[\"seqid\"],\n",
    "                row[\"source\"],\n",
    "                \"exon\",\n",
    "                str(exon[0]),\n",
    "                str(exon[1]),\n",
    "                row[\"score\"],\n",
    "                row[\"strand\"],\n",
    "                row[\"phase\"],\n",
    "                exon_attrs]\n",
    "        outFP.write(\"\\t\".join(elcs)+\"\\n\")\n",
    "\n",
    "    for cds in row[\"new_cds_chain\"]:\n",
    "        clcs = [row[\"seqid\"],\n",
    "                row[\"source\"],\n",
    "                \"CDS\",\n",
    "                str(cds[0]),\n",
    "                str(cds[1]),\n",
    "                row[\"score\"],\n",
    "                row[\"strand\"],\n",
    "                row[\"phase\"],\n",
    "                exon_attrs]\n",
    "        outFP.write(\"\\t\".join(clcs)+\"\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "a4d771f4-770f-418c-b535-2c3bea69c34d",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqid</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>attributes</th>\n",
       "      <th>tid</th>\n",
       "      <th>gid</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>exon_chain</th>\n",
       "      <th>cds_chain</th>\n",
       "      <th>has_cds</th>\n",
       "      <th>status</th>\n",
       "      <th>new_exon_chain</th>\n",
       "      <th>new_cds_chain</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr19</td>\n",
       "      <td>Liftoff</td>\n",
       "      <td>transcript</td>\n",
       "      <td>4460</td>\n",
       "      <td>6999</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"CHS.1.1\"; gene_id \"CHS.1\"; gene...</td>\n",
       "      <td>CHS.1.1</td>\n",
       "      <td>CHS.1</td>\n",
       "      <td>CHS.1</td>\n",
       "      <td>((4460, 4813), (5199, 5307), (5807, 6999))</td>\n",
       "      <td>()</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>((4460, 4813), (5199, 5307), (5807, 6999))</td>\n",
       "      <td>()</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>Liftoff</td>\n",
       "      <td>transcript</td>\n",
       "      <td>864517</td>\n",
       "      <td>868933</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"CHS.100.1\"; gene_id \"CHS.100\"; ...</td>\n",
       "      <td>CHS.100.1</td>\n",
       "      <td>CHS.100</td>\n",
       "      <td>CHS.100</td>\n",
       "      <td>((864517, 865022), (868548, 868933))</td>\n",
       "      <td>()</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>((864517, 865022), (868548, 868933))</td>\n",
       "      <td>()</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>Liftoff</td>\n",
       "      <td>transcript</td>\n",
       "      <td>864517</td>\n",
       "      <td>868933</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"CHS.100.2\"; gene_id \"CHS.100\"; ...</td>\n",
       "      <td>CHS.100.2</td>\n",
       "      <td>CHS.100</td>\n",
       "      <td>CHS.100</td>\n",
       "      <td>((864517, 865022), (868538, 868933))</td>\n",
       "      <td>()</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>((864517, 865022), (868538, 868933))</td>\n",
       "      <td>()</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>Liftoff</td>\n",
       "      <td>transcript</td>\n",
       "      <td>864517</td>\n",
       "      <td>868933</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"CHS.100.3\"; gene_id \"CHS.100\"; ...</td>\n",
       "      <td>CHS.100.3</td>\n",
       "      <td>CHS.100</td>\n",
       "      <td>CHS.100</td>\n",
       "      <td>((864517, 865314), (868538, 868933))</td>\n",
       "      <td>()</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>((864517, 865314), (868538, 868933))</td>\n",
       "      <td>()</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>CHESS</td>\n",
       "      <td>transcript</td>\n",
       "      <td>864517</td>\n",
       "      <td>869839</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>transcript_id \"CHS.100.8\"; gene_id \"CHS.100\"; ...</td>\n",
       "      <td>CHS.100.8</td>\n",
       "      <td>CHS.100</td>\n",
       "      <td>CHS.100</td>\n",
       "      <td>((864517, 865314), (868538, 868944), (869739, ...</td>\n",
       "      <td>()</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>((864517, 865314), (868538, 868944), (869739, ...</td>\n",
       "      <td>()</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   seqid   source        type   start     end score strand phase  \\\n",
       "0  chr19  Liftoff  transcript    4460    6999     .      +     .   \n",
       "1   chr1  Liftoff  transcript  864517  868933     .      -     .   \n",
       "2   chr1  Liftoff  transcript  864517  868933     .      -     .   \n",
       "3   chr1  Liftoff  transcript  864517  868933     .      -     .   \n",
       "4   chr1    CHESS  transcript  864517  869839     .      -     .   \n",
       "\n",
       "                                          attributes        tid      gid  \\\n",
       "0  transcript_id \"CHS.1.1\"; gene_id \"CHS.1\"; gene...    CHS.1.1    CHS.1   \n",
       "1  transcript_id \"CHS.100.1\"; gene_id \"CHS.100\"; ...  CHS.100.1  CHS.100   \n",
       "2  transcript_id \"CHS.100.2\"; gene_id \"CHS.100\"; ...  CHS.100.2  CHS.100   \n",
       "3  transcript_id \"CHS.100.3\"; gene_id \"CHS.100\"; ...  CHS.100.3  CHS.100   \n",
       "4  transcript_id \"CHS.100.8\"; gene_id \"CHS.100\"; ...  CHS.100.8  CHS.100   \n",
       "\n",
       "  gene_name                                         exon_chain cds_chain  \\\n",
       "0     CHS.1         ((4460, 4813), (5199, 5307), (5807, 6999))        ()   \n",
       "1   CHS.100               ((864517, 865022), (868548, 868933))        ()   \n",
       "2   CHS.100               ((864517, 865022), (868538, 868933))        ()   \n",
       "3   CHS.100               ((864517, 865314), (868538, 868933))        ()   \n",
       "4   CHS.100  ((864517, 865314), (868538, 868944), (869739, ...        ()   \n",
       "\n",
       "   has_cds status                                     new_exon_chain  \\\n",
       "0        0    NaN         ((4460, 4813), (5199, 5307), (5807, 6999))   \n",
       "1        0    NaN               ((864517, 865022), (868548, 868933))   \n",
       "2        0    NaN               ((864517, 865022), (868538, 868933))   \n",
       "3        0    NaN               ((864517, 865314), (868538, 868933))   \n",
       "4        0    NaN  ((864517, 865314), (868538, 868944), (869739, ...   \n",
       "\n",
       "  new_cds_chain  \n",
       "0            ()  \n",
       "1            ()  \n",
       "2            ()  \n",
       "3            ()  \n",
       "4            ()  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[[\"new_exon_chain\", \"new_cds_chain\"]] = df.apply(lambda row: pd.Series(fix_coordinates(row)), axis=1)\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "3da3c807-21b8-4859-a77b-4128b512a35b",
   "metadata": {},
   "outputs": [],
   "source": [
    "outFP = open(outbase+\".p1.gtf\",\"w+\")\n",
    "df.apply(lambda row: pd.Series(to_gtf(row,outFP)), axis=1)\n",
    "outFP.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "2ea1b38d-818b-4ba7-887d-9a3ece937629",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# assign phase with gffread\n",
    "cmd = [\"gffread\",\"-o\",outbase+\".p2.gtf\",\"-T\",outbase+\".p1.gtf\"]\n",
    "subprocess.call(cmd)\n",
    "\n",
    "# extract proteins again\n",
    "cmd = [\"gffread\",\"-g\",ref_fasta,\"-x\",outbase+\".p2.cds_nt.fa\",outbase+\".p2.gtf\"]\n",
    "subprocess.call(cmd)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "4a8023af-1507-4dff-9bec-756038ce7459",
   "metadata": {},
   "outputs": [],
   "source": [
    "# load into a dictionary\n",
    "stop_status = definitions.load_fasta_dict(outbase+\".p2.cds_nt.fa\")\n",
    "# iterate and replace sequence with True/False based on whether last character is \"*\"\n",
    "for tid,seq in stop_status.items():\n",
    "    if seq[-3:] in [\"TAG\",\"TAA\",\"TGA\"]:\n",
    "        stop_status[tid]=True\n",
    "    else:\n",
    "        stop_status[tid]=False"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "66cb67e8-441e-4657-9993-a17e658d5069",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# now merge these updated results to the dataframe and repeat the steps this time only correcting those that need to be\n",
    "df['status_p2'] = df['tid'].map(stop_status)\n",
    "\n",
    "\n",
    "# process line by line - remember strands\n",
    "def fix_coordinates_p2(row):\n",
    "    if not row[\"has_cds\"]:\n",
    "        return [row[\"exon_chain\"], row[\"cds_chain\"]]\n",
    "    if not definitions.clen(row[\"cds_chain\"])%3==0: # invalid CDS\n",
    "        return [row[\"exon_chain\"], row[\"cds_chain\"]]\n",
    "    if not definitions.cut(row[\"exon_chain\"],row[\"cds_chain\"][0][0],row[\"cds_chain\"][-1][1])==list(row[\"cds_chain\"]): # if the exon cahin can not be cut to the cds chain\n",
    "        return [row[\"exon_chain\"], row[\"cds_chain\"]]\n",
    "    if row[\"status\"]==True: # already has stop\n",
    "        return [row[\"exon_chain\"], row[\"cds_chain\"]]\n",
    "    if row[\"status_p2\"]==False: # correction didn't work - just skip\n",
    "        return [row[\"exon_chain\"], row[\"cds_chain\"]]\n",
    "    if row[\"exon_chain\"][0][0]<4: # too close to start\n",
    "        return [row[\"exon_chain\"], row[\"cds_chain\"]]\n",
    "        \n",
    "    # otherwise process the data\n",
    "    c1,c2 = extend_intervals(list(row[\"exon_chain\"]),list(row[\"cds_chain\"]),3,row[\"strand\"]==\"+\")\n",
    "    return [c1,c2]\n",
    "\n",
    "df[[\"new_exon_chain\", \"new_cds_chain\"]] = df.apply(lambda row: pd.Series(fix_coordinates_p2(row)), axis=1)\n",
    "\n",
    "outFP = open(outbase+\".corrected.gtf\",\"w+\")\n",
    "df.apply(lambda row: pd.Series(to_gtf(row,outFP)), axis=1)\n",
    "outFP.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "81bfb828-d0a7-4446-91fc-5303fead3ef8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cmd = [\"gffread\",\"-o\",outbase+\".corrected.gffread.gtf\",\"-T\",\"-F\",\"--keep-exon-attrs\",outbase+\".corrected.gtf\"]\n",
    "subprocess.call(cmd)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "0d213b4a-66bb-4c67-b236-60e63eb84766",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/ccb/salz8-1/avaraby/chess_maintenance_scripts/tmp/chess3.0.1.CHM13.fix_dup_gid.adj_stop.corrected.gffread.gtf'"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "os.path.abspath(outbase+\".corrected.gffread.gtf\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c3db58e4-74bb-41ab-bf9b-9b7235b1dc23",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
